
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Trajectory analysis &#8212; GSCN workshop 2024</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!-- 
    this give us a css class that will be invisible only if js is disabled 
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=26a4bc78f4c0ddb94549"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549" />

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'session-3.1';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Multiple samples" href="session-3.2.html" />
    <link rel="prev" title="Dimensionality reduction, clustering and annotation" href="session-2.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">GSCN workshop 2024</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    GSCN workshop 2024
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Session 1</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="session-1.html">Pre-processing</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Session 2</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="session-2.html">Dimensionality reduction, clustering and annotation</a></li>











</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Session 3</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Trajectory analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="session-3.2.html">Multiple samples</a></li>





</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/sarajimenez/GSCN_workshop_2024" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/session-3.1.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Trajectory analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Trajectory analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-packages">Load packages</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-processed-data">Load processed data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#trajectory-inference">Trajectory Inference</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#diffusion-maps">Diffusion Maps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-scanpy-inbuilt-diffusion-maps">Method 1: Scanpy inbuilt diffusion maps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-data-2-cells-along-a-continuum">Example data 2: Cells along a continuum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nearest-neighbor-and-diffusion-maps">Nearest neighbor and diffusion maps</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#palantir">Palantir</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-start-cell">Define start cell</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-terminal-states">Define terminal states</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-pseudotime">Compute pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#get-the-pseudotime">Get the pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-pseudotime">Visualize the pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-probability-for-cells-to-reach-certain-terminal-state">Visualize the probability for cells to reach certain terminal state</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-gene-expression-trends">Visualize gene expression trends</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clustering-gene-expression-trends">Clustering gene expression trends</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#paga">PAGA</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paga-with-distances-computed-on-pca">PAGA with distances computed on PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paga-on-diffusion-components">PAGA on diffusion components</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#using-results-to-quantify-distance-between-clusters">Using results to quantify distance between clusters</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="trajectory-analysis">
<h1>Trajectory analysis<a class="headerlink" href="#trajectory-analysis" title="Link to this heading">#</a></h1>
<p>In this notebook, we will explore the problem of trajectory analysis</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="load-packages">
<h1>Load packages<a class="headerlink" href="#load-packages" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="load-processed-data">
<h1>Load processed data<a class="headerlink" href="#load-processed-data" title="Link to this heading">#</a></h1>
<p>We will continue with the data used in session-2:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_path</span> <span class="o">=</span> <span class="s1">&#39;/Users/sara.jimenez/Documents/scWorkshop/data/&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_clean</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;processed/annotated_data.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_clean</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 11828 × 17698
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;mt_pct_content&#39;, &#39;ribo_pct_content&#39;, &#39;doublet_score&#39;, &#39;doublet&#39;, &#39;G2_M_checkpoint&#39;, &#39;leiden_r1.5&#39;, &#39;leiden_r0.5&#39;, &#39;leiden&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;leiden_r0.5_colors&#39;, &#39;leiden_r1.5_colors&#39;, &#39;log1p&#39;, &#39;neighbors_30&#39;, &#39;on_norm_log&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_data&#39;
    obsp: &#39;neighbors_30_connectivities&#39;, &#39;neighbors_30_distances&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="trajectory-inference">
<h1>Trajectory Inference<a class="headerlink" href="#trajectory-inference" title="Link to this heading">#</a></h1>
<p>scRNA-seq has been applied extensively to study developmental systems. It is now generally understood that the process of differentiation is a continuous process rather than a series of discreet/distinct stages. Trajectory inference is a computational framework to align the cells developing from one phenotype to another along a continuous axis (also called pseudotime).</p>
<p>Underlying pseudotime analysis is the assumption that because differentiation process is asynchronous, a single snapshot of scRNA-seq data comprises of cells in all possible states during the process. This allows us to utilize the heterogeneity among the cells to align them along a continuous axis. In essence, we are interested in capturing gradual changes during the process, which means we ought to understand and map local similarity between cells. As we have discussed extensively in sessions 2 and 3, graphs are the perfect mathematical tools to do so. Therefore, all the methods we discuss below will rely on modeling the data using a graph.</p>
<p>There are tens of algorithms developed to perform pseudotime analysis on single-cell RNA-seq data. Here, we will discuss two important methods, namely diffusion maps and Palantir, for their effectiveness and success we have seen in scRNA-seq data analysis. ( Note: Palantir develops on top of diffusion maps).</p>
<p>Refs:</p>
<p><a class="reference external" href="https://www.nature.com/articles/s41587-019-0071-9?platform=hootsuite">https://www.nature.com/articles/s41587-019-0071-9?platform=hootsuite</a></p>
<p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/30899105/">https://pubmed.ncbi.nlm.nih.gov/30899105/</a></p>
<p><a class="reference external" href="https://www.nature.com/articles/nmeth.3971">https://www.nature.com/articles/nmeth.3971</a></p>
<section id="diffusion-maps">
<h2>Diffusion Maps<a class="headerlink" href="#diffusion-maps" title="Link to this heading">#</a></h2>
<p>Diffusion Maps are a suite of graph-based techniques to perform non-linear dimensionality reduction of high-dimensional data. They can be crudely thought of as non-linear version of PCA: just like PCA, diffusion maps will identify axis of high variances in the data, but unlike PCA - which is linear - diffusion maps take into consideration  non-linear associations between features (genes in our case) of the data. In other words, diffusion maps transform any high-dimensional data into a lower dimensional Euclidean space by accounting for any non-linear dependencies between the features. In doing so, it preserves the relationship between points that are similar in high-dimensional space and discards distant relationships.</p>
<p>This property of diffusion maps makes it very favorable for single-cell data analysis. By preserving the local neighborhood structure, it provides a platform to incrementally navigate along the data manifold. Unlike diffusion maps, UMAP (and Force Directed Layout) - while useful to visualize data - does not accurately preserve high-dimensional pairwise distances in the resulting low-dimensional representation. This limits the use of UMAP (and Force Directed Layout) to visualization and does not facilitate any further quantitative analysis. However, since diffusion maps are distance preserving, they can be used for downstream analysis.</p>
<p>The need for using diffusion maps also stems from our assumption that scRNA-seq data are inherently noisy and biological signals are masked by the noise. As such, we typically begin by performing principal component analysis as a way to decouple strong variability in the data (true biological signal) from random noise. As we discussed in our previous workshop, this is achieved by retrieving only the top <code class="docutils literal notranslate"><span class="pre">p</span></code> principal components and neglecting the rest (which typically identify with noise in the data) for any downstream analysis. However, we have observed with scRNA-seq data that PCA is generally ideal for working with global aspects of the data (such as to cluster the cells on, visualize the data etc.) but is not <em>always</em> optimal for a finer analysis such as pseudotime alignment of individual cells.</p>
<p>One of the popular ways to study cells along a continuum of gene expression space (as opposed to distinct clusters) is to apply diffusion maps on the retrieved principal components. Because the resulting embedding aims to preserve the local neighborhood of each cell, the diffusion maps form a natural coordinate system for cells along a continuum. As a result, they have been most successfully employed to align cells undergoing development along a trajectory. The success of diffusion maps relies on its ability to capture the geometry of the data accurately.</p>
<p>Diffusion maps work by first building a k-nearest neighbor graph on the principal components (similar to clustering analysis, see session 1 or previous workshop material), followed by conversion of the distance graph to an affinity graph. While there are various models/methods proposed to convert distance matrix to affinity matrix, we recommend using adaptive Gaussian kernel. The kernel ensures that affinity between cells decrease exponentially with distance between cells in every direction (of expression or PCA space) but remains sensitive to density of cells near every cell. In other words, if the neighborhood of a cell is densely sampled then the similarity between cells is more concentrated while if the neighborhood of a cell is sparsely sampled then the similarity is more diffuse. This is important to avoid bias in analysis due to sampling differences in a heterogeneous population of cells. The affinity matrix is then converted into a Markov matrix, which represents a probabilistic framework of cellular state neighborhood. To elaborate, for a data of K cells, the Markov matrix is of size K-by-K; each row of the matrix sums to 1 and the (i, j)-th entry of the matrix represents the probability of a cell transitioning from state i to state j in one time step. Therefore, the Markov matrix automatically provides a probabilistic representation of cellular transition and is well-suited for pseudotime analysis that aims to align the cells along their differentiation transition states. Furthermore, because we begin with the k-nearest neighbor graph, only the immediate neighborhood each cell (i.e. phenotypically most similar cells) get high probability score of transition. Once the Markov matrix is constructed, we obtain the eigenvalues and eigenvectors of the matrix (analogous to the procedure in PCA) and order the eigenvectors based on the associated eigenvalue. We interpret the top eigenvector (we ignore the first eigenvector with eigenvalue of 1, see references below for this) as the direction of highest variability in the data, and often can be used to study cellular transitions in the data. These eigenvectors are called the diffusion components or diffusion maps (maps because they map the data from high-dimensional gene expression or PCA space to a lower dimensional diffusion space). Similarly, we encourage you to explore the next few eigenvectors to study any potential axes of continuum in the data. We will discuss more below.</p>
<p>You will notice that this is very similar to how we performed MAGIC above, and indeed the construction of Markov Matrix is identical and it is used for two separate scenarios.</p>
<p>Refs:
<a class="reference external" href="https://www.nature.com/articles/nmeth.3971">https://www.nature.com/articles/nmeth.3971</a></p>
<p><a class="reference external" href="https://www.nature.com/articles/nbt.3569">https://www.nature.com/articles/nbt.3569</a></p>
<p><a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S1063520306000546">https://www.sciencedirect.com/science/article/pii/S1063520306000546</a></p>
<p>The size of the nearest neighborhood (<code class="docutils literal notranslate"><span class="pre">k</span></code> or <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> parameter) is a crucial parameter and analogous to clustering analysis, we recommend you to choose it such that it is small enough to only consider local neighbors of each cell, but large enough that you do not get discrete sub-graphs. A typical choice of <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> is <code class="docutils literal notranslate"><span class="pre">30</span></code>. Furthermore, analogous to clustering analysis, it is highly recommended that you show your results do not vary with slight alterations in the value of <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code>. For example, if you wish to choose <code class="docutils literal notranslate"><span class="pre">n_neighbors</span> <span class="pre">=</span> <span class="pre">30</span></code> then we recommend that you recompute the graph for various values of <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> around <code class="docutils literal notranslate"><span class="pre">30</span></code> such as <code class="docutils literal notranslate"><span class="pre">n_neighbors</span> <span class="pre">=</span> <span class="pre">(21,</span> <span class="pre">24,</span> <span class="pre">27,</span> <span class="pre">30,</span> <span class="pre">33,</span> <span class="pre">36,</span> <span class="pre">39)</span></code> and show that the meaningful diffusion components you obtain for all these values of <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> are highly correlated to each other.</p>
<section id="method-1-scanpy-inbuilt-diffusion-maps">
<h3>Method 1: Scanpy inbuilt diffusion maps<a class="headerlink" href="#method-1-scanpy-inbuilt-diffusion-maps" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_clean</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 11828 × 17698
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;mt_pct_content&#39;, &#39;ribo_pct_content&#39;, &#39;doublet_score&#39;, &#39;doublet&#39;, &#39;G2_M_checkpoint&#39;, &#39;leiden_r1.5&#39;, &#39;leiden_r0.5&#39;, &#39;leiden&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;leiden_r0.5_colors&#39;, &#39;leiden_r1.5_colors&#39;, &#39;log1p&#39;, &#39;neighbors_30&#39;, &#39;on_norm_log&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_data&#39;
    obsp: &#39;neighbors_30_connectivities&#39;, &#39;neighbors_30_distances&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># n_comps is the number of top eigenvalues/eigenvectors you would like to compute, 20-30 is usually sufficient</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">diffmap</span><span class="p">(</span><span class="n">adata_clean</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_clean</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 11828 × 17698
    obs: &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;mt_pct_content&#39;, &#39;ribo_pct_content&#39;, &#39;doublet_score&#39;, &#39;doublet&#39;, &#39;G2_M_checkpoint&#39;, &#39;leiden_r1.5&#39;, &#39;leiden_r0.5&#39;, &#39;leiden&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;leiden_r0.5_colors&#39;, &#39;leiden_r1.5_colors&#39;, &#39;log1p&#39;, &#39;neighbors_30&#39;, &#39;on_norm_log&#39;, &#39;pca&#39;, &#39;rank_genes_groups&#39;, &#39;umap&#39;, &#39;diffmap_evals&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_diffmap&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_data&#39;
    obsp: &#39;neighbors_30_connectivities&#39;, &#39;neighbors_30_distances&#39;
</pre></div>
</div>
</div>
</div>
<p>We can view the eigenvalues as below. The first eigenvalue is exactly 1 and this expected due to a property of any Markov matrix (any matrix whose rows sum to 1). This only reflects that there is one big connected component in the data (which is a common property across any data by construction and not specific to the current data) and we ignore this eigenvalue and the corresponding eigenvector. Note: if you see multiple eigenvalues exactly equal to 1, which is rare in scRNA-seq data specifically in the context of development, then that has a different, and important, interpretation. Please see the references above.</p>
<p>We interpret the next eigenvalue and the associated eigenvector as the most informative component about the data that indicates direction of most biological variability. Often this indicates towards a possible transition between cell states.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_clean</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])),</span> <span class="n">adata_clean</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="s1">&#39;on&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Index&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Eigen-Values&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;Eigen-Values&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/321219898f58495b6c4001aa4c25f4be614f7b18b6a6f6312c188595c7176977.png"><img alt="_images/321219898f58495b6c4001aa4c25f4be614f7b18b6a6f6312c188595c7176977.png" src="_images/321219898f58495b6c4001aa4c25f4be614f7b18b6a6f6312c188595c7176977.png" style="width: 572px; height: 422px;" /></a>
</div>
</div>
<p>We also use the first large gap in the eigenvalue spectrum (i.e. the plot above) to quantify the number of most meaningful diffusion components to consider for analysis. For example, in the plot above, we see a substantial gap between 6th and 7th eigenvalue. This indicates that perhaps the first 6 eigenvalues and associated eigenvectors are the most meaningful. This may seem a bit arbitrary (especially when you see that there is a much bigger change between 17th and 18th eigenvector), and we will discuss how to avoid this pitfall in the next session (see Palantir).</p>
<p>We visualize the top 10 eigenvectors as follows (Note: we must ignore the first i.e. 0th eigenvector from downstream analysis):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ncol</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">nrow</span> <span class="o">=</span> <span class="mi">3</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="n">ncol</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">nrow</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">11</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">nrow</span><span class="p">,</span> <span class="n">ncol</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_clean</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_clean</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
                 <span class="n">c</span> <span class="o">=</span> <span class="n">adata_clean</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Diff Map - &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>
    
<span class="c1"># To save: </span>
<span class="c1"># fig.savefig(output_directory + &#39;umap_colored_by_diff_maps.png&#39;, bbox_inches = &#39;tight&#39;, dpi = 150)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/342411c0bbfd7f2dae69a387f1cdbb350e26b028a9e46db54f8f354f2bc281c6.png"><img alt="_images/342411c0bbfd7f2dae69a387f1cdbb350e26b028a9e46db54f8f354f2bc281c6.png" src="_images/342411c0bbfd7f2dae69a387f1cdbb350e26b028a9e46db54f8f354f2bc281c6.png" style="width: 2010px; height: 1132px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_clean</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/3f70383f277e102a33c0a5efeb3e1b346a5f5065b9d59cea300b8ec60fc32ca7.png"><img alt="_images/3f70383f277e102a33c0a5efeb3e1b346a5f5065b9d59cea300b8ec60fc32ca7.png" src="_images/3f70383f277e102a33c0a5efeb3e1b346a5f5065b9d59cea300b8ec60fc32ca7.png" style="width: 631px; height: 417px;" /></a>
</div>
</div>
<p>Interpreting the diffusion maps:</p>
<ol class="arabic simple">
<li><p>We ignore the 0th diffusion component (for reasons, please see the text above and the provided references).</p></li>
<li><p>The next top diffusion component (Diff Map - 1) separates out clusters 6, 0 and 1.</p></li>
<li><p>Diff Map - 2 separates out clusters 4, 2, 3, 5, and 9.</p></li>
<li><p>Diff Map - 3 identifies the small cluster 9 as being distinct from rest of the data. Note: if C9 is so different and is not clear what kind of cell type it is, diffusion maps could be hinting that it is a rare and distinct or a noisy cluster of cells. As such, diffusion maps can be used for data clean-up as well.</p></li>
</ol>
<p>And so forth.</p>
<p>To summarize, diffusion maps attempt to highlight the differences or variablity in the data. However, we set out to illustrate the use of diffusion maps for the context of aligning cells on a continuous axis and as such, applying it to data with clear separated clusters (like the example above) is not ideal. Therefore, we will run diffusion maps on a separate data (provided along with this notebook) and interpret the findings.</p>
</section>
<section id="example-data-2-cells-along-a-continuum">
<h3>Example data 2: Cells along a continuum<a class="headerlink" href="#example-data-2-cells-along-a-continuum" title="Link to this heading">#</a></h3>
<p>The data is taken from <a class="reference external" href="https://www.nature.com/articles/s41587-019-0068-4">https://www.nature.com/articles/s41587-019-0068-4</a> and downloaded from <a class="github reference external" href="https://github.com/dpeerlab/Palantir">dpeerlab/Palantir</a>. The data has already been normalized, clustered, cell type annotated and visualization maps have also been computed.</p>
<p>The data consists of CD34+ human bone marrow cells undergoing differentiation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;/human_cd34_bm_rep1.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Change the name of the obsm so we can invoke scanpy plotting functions</span>
<span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;clusters&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/c63b43877c5aceecb42ff141fb31634215bb124d81a3fa8d59b923d175e5cc24.png"><img alt="_images/c63b43877c5aceecb42ff141fb31634215bb124d81a3fa8d59b923d175e5cc24.png" src="_images/c63b43877c5aceecb42ff141fb31634215bb124d81a3fa8d59b923d175e5cc24.png" style="width: 573px; height: 417px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize some gene expression trends</span>
<span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD34&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF8&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79A&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGA2B&#39;</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">genes</span><span class="p">,</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="s1">&#39;p1&#39;</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="s1">&#39;p99&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/c965565215dfe90336123cb96ad0d7bf18fa2af73decbf32e26ef340986fbf2b.png"><img alt="_images/c965565215dfe90336123cb96ad0d7bf18fa2af73decbf32e26ef340986fbf2b.png" src="_images/c965565215dfe90336123cb96ad0d7bf18fa2af73decbf32e26ef340986fbf2b.png" style="width: 1733px; height: 840px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We are computing 300 principal components as that is what the authors used. </span>
<span class="c1"># The rationale was that they wanted to consider enough PCs to explain 75% of the variance in the data</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_comps</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;
    varm: &#39;PCs&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="nearest-neighbor-and-diffusion-maps">
<h3>Nearest neighbor and diffusion maps<a class="headerlink" href="#nearest-neighbor-and-diffusion-maps" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute diffusion maps</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span> <span class="o">=</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">diffmap</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Eigenvalues</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])),</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="s1">&#39;on&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Index&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Eigen-Values&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;Eigen-Values&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/48a1f37dfa496fd93405f71c800079c64b8f43a48a513f3e875097998ff19174.png"><img alt="_images/48a1f37dfa496fd93405f71c800079c64b8f43a48a513f3e875097998ff19174.png" src="_images/48a1f37dfa496fd93405f71c800079c64b8f43a48a513f3e875097998ff19174.png" style="width: 573px; height: 422px;" /></a>
</div>
</div>
<p>We see a clear gap between 8th and 9th eigenvalues (note: counting starts at 0). This is indicative that the first 8 eigenvalues are the most informative about the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the first 12 eigenvectors</span>
<span class="n">ncol</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">nrow</span> <span class="o">=</span> <span class="mi">3</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="n">ncol</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">nrow</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">12</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">nrow</span><span class="p">,</span> <span class="n">ncol</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
                 <span class="n">c</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Diff Map - &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>
    
<span class="c1"># To save: </span>
<span class="c1"># fig.savefig(output_directory + &#39;umap_colored_by_diff_maps.png&#39;, bbox_inches = &#39;tight&#39;, dpi = 150)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/2d18ee0ddf8e2f186318d67d79c1cb375016f612dab3d650c7600a42a824fa20.png"><img alt="_images/2d18ee0ddf8e2f186318d67d79c1cb375016f612dab3d650c7600a42a824fa20.png" src="_images/2d18ee0ddf8e2f186318d67d79c1cb375016f612dab3d650c7600a42a824fa20.png" style="width: 2010px; height: 1128px;" /></a>
</div>
</div>
<p>Interpreting the plots above:</p>
<ol class="arabic simple">
<li><p>We ignore Diff Map - 0 as we have discussed previously.</p></li>
<li><p>Diff Map - 1 tracks the progression of GATA1+ cells (Erythroid cells)</p></li>
<li><p>Diff Map - 2 tracks the progression of CD79A+ cells (Common Lymphoid Progrenitor cells)</p></li>
<li><p>Diff Map - 3 tracks the progression of IRF8+ and MPO+ cells (Dendritic cells and Monocytes)</p></li>
<li><p>Diff Map - 4 trakcs the progression of IRF8+ cells and distinguishes it from MPO+ cells (contrast with Diff Map -3)</p></li>
</ol>
<p>and so forth.</p>
<p>As we can see, diffusion maps clearly track the major changes along the data. The above is a complex example of multiple trajectories or transitions but is a good example case that illustrates the power of diffusion maps in delineating such trajectories.</p>
<p>To perform downstream analysis or interpreting the diffusion components, you can start to identify genes that are correlated (positively or negatively) with the diffusion component that you are interested in investigating. For example, we will correlate the top 10 diffusion components above with every gene and sort them based on their correlation score. First, we will provide code to correlate one diffusion component. Second, we will write a for loop to iterate it over all diffusion components of interest.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Correlate with diffusion component 1 (Recall: Diff Comp 0 is to be ignored)</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">X</span>
<span class="n">diff_1</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">]</span>

<span class="n">corr_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
    <span class="n">corr_temp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">diff_1</span><span class="p">)</span>
    <span class="n">corr_score</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">corr_temp</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Arrange that in pandas data frame with associated gene names:</span>
<span class="n">df_corr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;corr_score&#39;</span><span class="p">:</span> <span class="n">corr_score</span><span class="p">},</span> <span class="n">index</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">df_corr</span><span class="p">[</span><span class="s1">&#39;absolute_corr_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_corr</span><span class="p">[</span><span class="s1">&#39;corr_score&#39;</span><span class="p">])</span>
<span class="n">df_corr</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>corr_score</th>
      <th>absolute_corr_score</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A1BG</th>
      <td>-0.167827</td>
      <td>0.167827</td>
    </tr>
    <tr>
      <th>A2M</th>
      <td>-0.003106</td>
      <td>0.003106</td>
    </tr>
    <tr>
      <th>A2ML1</th>
      <td>0.044155</td>
      <td>0.044155</td>
    </tr>
    <tr>
      <th>A4GALT</th>
      <td>0.326927</td>
      <td>0.326927</td>
    </tr>
    <tr>
      <th>AAAS</th>
      <td>0.156086</td>
      <td>0.156086</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort by absolute value of corr_score (to isolate top positively &amp; negatively correlated genes)</span>
<span class="n">df_corr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;absolute_corr_score&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>corr_score</th>
      <th>absolute_corr_score</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>KLF1</th>
      <td>0.824020</td>
      <td>0.824020</td>
    </tr>
    <tr>
      <th>APOC1</th>
      <td>0.793446</td>
      <td>0.793446</td>
    </tr>
    <tr>
      <th>CNRIP1</th>
      <td>0.778498</td>
      <td>0.778498</td>
    </tr>
    <tr>
      <th>GATA1</th>
      <td>0.762808</td>
      <td>0.762808</td>
    </tr>
    <tr>
      <th>ANK1</th>
      <td>0.748084</td>
      <td>0.748084</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>SIRPD</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>SLC7A10</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>SNCG</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>TM4SF18</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>WLS</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>14651 rows × 2 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the trend of top 5 genes along diffusion component 1:</span>
<span class="n">top_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KLF1&#39;</span><span class="p">,</span> <span class="s1">&#39;APOC1&#39;</span><span class="p">,</span> <span class="s1">&#39;CNRIP1&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">,</span> <span class="s1">&#39;ANK1&#39;</span><span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">top_genes</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> 
               <span class="n">adata_bm</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">gene</span><span class="p">)],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Diffusion component-1&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/c95d5f6a41e4581f83acddcb48fd77752761ef94f7cb34b79fe08aca4f3d2e19.png"><img alt="_images/c95d5f6a41e4581f83acddcb48fd77752761ef94f7cb34b79fe08aca4f3d2e19.png" src="_images/c95d5f6a41e4581f83acddcb48fd77752761ef94f7cb34b79fe08aca4f3d2e19.png" style="width: 1558px; height: 786px;" /></a>
</div>
</div>
<p>As we can see, KLF1, GATA1 are some of the highly correlated genes with diffusion component 1. Inspecting the top genes this way can help you get a sense of what each diffusion component is trying to explain.</p>
<p>We can now automate this for all the diffusion components of interest:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Correlate with top 10 diffusion components</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">X</span>

<span class="k">for</span> <span class="n">diff_comp</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">diff_comp</span><span class="p">)</span>
    <span class="n">diff_temp</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="n">diff_comp</span><span class="p">]</span>
    <span class="n">corr_score</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="n">corr_temp</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">corrcoef</span><span class="p">(</span><span class="n">data</span><span class="p">[:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">diff_temp</span><span class="p">)</span>
        <span class="n">corr_score</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">corr_temp</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
    
    <span class="n">df_corr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;corr_score&#39;</span><span class="p">:</span> <span class="n">corr_score</span><span class="p">},</span> <span class="n">index</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">df_corr</span><span class="p">[</span><span class="s1">&#39;absolute_corr_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">df_corr</span><span class="p">[</span><span class="s1">&#39;corr_score&#39;</span><span class="p">])</span>
    
    <span class="c1"># save as csv:</span>
    <span class="c1">#df_corr.to_csv(output_directory + &#39;gene_correlation_with_diffusion_comp_&#39; + str(diff_comp) + &#39;.csv&#39;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>3
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>4
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>6
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>7
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>8
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>9
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>10
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2854: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/numpy/lib/function_base.py:2855: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="palantir">
<h2>Palantir<a class="headerlink" href="#palantir" title="Link to this heading">#</a></h2>
<p>Palantir is a graph-based method to align differentiating cells onto a pseudotime axis. Palantir models the differentiation landscape as a continuum (as opposed to a set of discrete stages) and aligns cells along an axis, automatically identifies and characterizes the terminal differentiated states, and assigns each cell a probablity of differentiating into one of the terminal states.</p>
<p>The key inputs to the method are:</p>
<ol class="arabic simple">
<li><p>scRNA-seq data from a differentiating system</p></li>
<li><p>Start cell, i.e. where the differentiation starts</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">k</span></code> or <code class="docutils literal notranslate"><span class="pre">n_neighbors</span></code> for graph construction.</p></li>
</ol>
<p>We again recommend that you investigate and ensure that your pseudotime and branch probability results are robust to slight variations around input 2 and 3. You can consult Supplementary Figures 5-8 in the reference below:</p>
<p>Ref: <a class="reference external" href="https://www.nature.com/articles/s41587-019-0068-4">https://www.nature.com/articles/s41587-019-0068-4</a></p>
<p>While palantir is integrated into scanpy as <code class="docutils literal notranslate"><span class="pre">sc.external.tl.palantir</span></code>, we suggest you import palantir by itself. The authors have made modifications for easier downstream analysis, which may not be up to date in Scanpy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>As part of the analysis, palantir first computes diffusion maps on the data, then scales the diffusion components based on their eigenvalues (please see the manuscript for more details). Then it uses the rescaled diffusion components to build the pseudotime.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>palantir
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Collecting palantir
  Downloading palantir-1.3.3-py3-none-any.whl (50 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">50.5/50.5 kB</span> <span class=" -Color -Color-Red">1.3 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25hRequirement already satisfied: pandas&gt;=0.22.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.5.3)
Requirement already satisfied: scikit-learn in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.2.2)
Collecting mellon&gt;=1.3.0
  Downloading mellon-1.4.3-py3-none-any.whl (95 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">96.0/96.0 kB</span> <span class=" -Color -Color-Red">2.7 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25hRequirement already satisfied: scipy&gt;=1.3 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.10.1)
Requirement already satisfied: joblib in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.2.0)
Requirement already satisfied: numpy&gt;=1.14.2 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.24.3)
Requirement already satisfied: anndata&gt;=0.8.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (0.8.0)
Collecting pygam
  Downloading pygam-0.9.1-py3-none-any.whl (522 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">522.0/522.0 kB</span> <span class=" -Color -Color-Red">2.2 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>a <span class=" -Color -Color-Cyan">0:00:01</span>
?25hCollecting fcsparser&gt;=0.1.2
  Downloading fcsparser-0.2.8-py3-none-any.whl (14.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">14.6/14.6 MB</span> <span class=" -Color -Color-Red">5.5 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>00:0100:01
?25hRequirement already satisfied: scanpy&gt;=1.6.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (1.9.3)
Requirement already satisfied: matplotlib&gt;=2.2.2 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (3.7.1)
Requirement already satisfied: networkx&gt;=2.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (3.1)
Requirement already satisfied: leidenalg&gt;=0.9.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from palantir) (0.9.1)
Requirement already satisfied: packaging&gt;=20 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from anndata&gt;=0.8.0-&gt;palantir) (23.1)
Requirement already satisfied: h5py&gt;=3 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from anndata&gt;=0.8.0-&gt;palantir) (3.8.0)
Requirement already satisfied: natsort in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from anndata&gt;=0.8.0-&gt;palantir) (8.3.1)
Requirement already satisfied: igraph&lt;0.11,&gt;=0.10.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from leidenalg&gt;=0.9.1-&gt;palantir) (0.10.4)
Requirement already satisfied: kiwisolver&gt;=1.0.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (1.4.4)
Requirement already satisfied: fonttools&gt;=4.22.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (4.39.4)
Requirement already satisfied: cycler&gt;=0.10 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (0.11.0)
Requirement already satisfied: pillow&gt;=6.2.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (9.5.0)
Requirement already satisfied: python-dateutil&gt;=2.7 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (2.8.2)
Requirement already satisfied: contourpy&gt;=1.0.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (1.0.7)
Requirement already satisfied: pyparsing&gt;=2.3.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (3.0.9)
Requirement already satisfied: importlib-resources&gt;=3.2.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from matplotlib&gt;=2.2.2-&gt;palantir) (5.12.0)
Requirement already satisfied: jax in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from mellon&gt;=1.3.0-&gt;palantir) (0.4.10)
Collecting jaxopt
  Downloading jaxopt-0.8.3-py3-none-any.whl (172 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">172.3/172.3 kB</span> <span class=" -Color -Color-Red">6.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>a <span class=" -Color -Color-Cyan">0:00:01</span>
?25hRequirement already satisfied: pytz&gt;=2020.1 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from pandas&gt;=0.22.0-&gt;palantir) (2023.3)
Requirement already satisfied: statsmodels&gt;=0.10.0rc2 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (0.14.0)
Requirement already satisfied: patsy in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (0.5.3)
Requirement already satisfied: session-info in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (1.0.0)
Requirement already satisfied: numba&gt;=0.41.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (0.57.0)
Requirement already satisfied: umap-learn&gt;=0.3.10 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (0.5.3)
Requirement already satisfied: seaborn in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (0.12.2)
Requirement already satisfied: tqdm in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scanpy&gt;=1.6.0-&gt;palantir) (4.65.0)
Requirement already satisfied: threadpoolctl&gt;=2.0.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from scikit-learn-&gt;palantir) (3.1.0)
Collecting scipy&gt;=1.3
  Downloading scipy-1.11.4-cp39-cp39-macosx_12_0_arm64.whl (29.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">29.7/29.7 MB</span> <span class=" -Color -Color-Red">4.1 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>00:0100:01
?25hCollecting numpy&gt;=1.14.2
  Downloading numpy-1.26.4-cp39-cp39-macosx_11_0_arm64.whl (14.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">14.0/14.0 MB</span> <span class=" -Color -Color-Red">3.9 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>00:0100:01
?25hCollecting progressbar2&lt;5.0.0,&gt;=4.2.0
  Downloading progressbar2-4.5.0-py3-none-any.whl (57 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">57.1/57.1 kB</span> <span class=" -Color -Color-Red">3.6 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>
?25hRequirement already satisfied: texttable&gt;=1.6.2 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from igraph&lt;0.11,&gt;=0.10.0-&gt;leidenalg&gt;=0.9.1-&gt;palantir) (1.6.7)
Requirement already satisfied: zipp&gt;=3.1.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from importlib-resources&gt;=3.2.0-&gt;matplotlib&gt;=2.2.2-&gt;palantir) (3.15.0)
Requirement already satisfied: llvmlite&lt;0.41,&gt;=0.40.0dev0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from numba&gt;=0.41.0-&gt;scanpy&gt;=1.6.0-&gt;palantir) (0.40.0)
Collecting numba&gt;=0.41.0
  Downloading numba-0.60.0-cp39-cp39-macosx_11_0_arm64.whl (2.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">2.7/2.7 MB</span> <span class=" -Color -Color-Red">4.0 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>00:0100:01
?25hCollecting llvmlite&lt;0.44,&gt;=0.43.0dev0
  Downloading llvmlite-0.43.0-cp39-cp39-macosx_11_0_arm64.whl (28.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ <span class=" -Color -Color-Green">28.8/28.8 MB</span> <span class=" -Color -Color-Red">10.4 MB/s</span> eta <span class=" -Color -Color-Cyan">0:00:00</span>00:0100:01
?25hCollecting python-utils&gt;=3.8.1
  Downloading python_utils-3.9.0-py2.py3-none-any.whl (32 kB)
Requirement already satisfied: six&gt;=1.5 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from python-dateutil&gt;=2.7-&gt;matplotlib&gt;=2.2.2-&gt;palantir) (1.16.0)
Requirement already satisfied: pynndescent&gt;=0.5 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from umap-learn&gt;=0.3.10-&gt;scanpy&gt;=1.6.0-&gt;palantir) (0.5.10)
Requirement already satisfied: ml-dtypes&gt;=0.1.0 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from jax-&gt;mellon&gt;=1.3.0-&gt;palantir) (0.1.0)
Requirement already satisfied: opt-einsum in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from jax-&gt;mellon&gt;=1.3.0-&gt;palantir) (3.3.0)
Requirement already satisfied: jaxlib&gt;=0.1.69 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from jaxopt-&gt;mellon&gt;=1.3.0-&gt;palantir) (0.4.10)
Requirement already satisfied: stdlib-list in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from session-info-&gt;scanpy&gt;=1.6.0-&gt;palantir) (0.8.0)
Requirement already satisfied: typing-extensions&gt;3.10.0.2 in /Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages (from python-utils&gt;=3.8.1-&gt;progressbar2&lt;5.0.0,&gt;=4.2.0-&gt;pygam-&gt;palantir) (4.6.1)
Installing collected packages: python-utils, numpy, llvmlite, scipy, progressbar2, numba, pygam, fcsparser, jaxopt, mellon, palantir
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.3
    Uninstalling numpy-1.24.3:
      Successfully uninstalled numpy-1.24.3
  Attempting uninstall: llvmlite
    Found existing installation: llvmlite 0.40.0
    Uninstalling llvmlite-0.40.0:
      Successfully uninstalled llvmlite-0.40.0
  Attempting uninstall: scipy
    Found existing installation: scipy 1.10.1
    Uninstalling scipy-1.10.1:
      Successfully uninstalled scipy-1.10.1
  Attempting uninstall: numba
    Found existing installation: numba 0.57.0
    Uninstalling numba-0.57.0:
      Successfully uninstalled numba-0.57.0
Successfully installed fcsparser-0.2.8 jaxopt-0.8.3 llvmlite-0.43.0 mellon-1.4.3 numba-0.60.0 numpy-1.26.4 palantir-1.3.3 progressbar2-4.5.0 pygam-0.9.1 python-utils-3.9.0 scipy-1.11.4
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">palantir</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># first compute diffusion maps</span>
<span class="n">palantir</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">run_diffusion_maps</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">20</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualize the eigenvalues:</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;DM_EigenValues&#39;</span><span class="p">])),</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;DM_EigenValues&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Index&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;EigenValues&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="s1">&#39;on&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/e221c4b6b221c5c5cde6eea5f11379c9808d6545ce2d67657352f626812dbb7a.png"><img alt="_images/e221c4b6b221c5c5cde6eea5f11379c9808d6545ce2d67657352f626812dbb7a.png" src="_images/e221c4b6b221c5c5cde6eea5f11379c9808d6545ce2d67657352f626812dbb7a.png" style="width: 392px; height: 375px;" /></a>
</div>
</div>
<p>Like before we see a gap between 8th and 9th eigenvalues. Now, palantir will use this result to rescale the top N eigenvectors, which are chosen automatically (but can also be userdefined) based on the gap between eigenvalues as shown above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palantir</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">determine_multiscale_space</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors_multiscaled&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(5780, 7)
</pre></div>
</div>
</div>
</div>
<p>As we suspected, palantir considered the first 8 eigenvectors (and it removes the 0th eigenvector) as the most meaningful and rescaled them based on their eigenvalues (for more details, please see the manuscript).</p>
<p>Color tSNE by top 8 diffusion maps:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the first 8 eigenvectors</span>
<span class="n">ncol</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">nrow</span> <span class="o">=</span> <span class="mi">3</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="n">ncol</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">nrow</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">8</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">nrow</span><span class="p">,</span> <span class="n">ncol</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
                 <span class="n">c</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Diff Map - &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">))</span>
    
<span class="c1"># To save: </span>
<span class="c1"># fig.savefig(output_directory + &#39;umap_colored_by_diff_maps.png&#39;, bbox_inches = &#39;tight&#39;, dpi = 150)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/29e927b5c425529eab774093f6a8f9dba034e7220af62cd3bf6cefda1bef6f83.png"><img alt="_images/29e927b5c425529eab774093f6a8f9dba034e7220af62cd3bf6cefda1bef6f83.png" src="_images/29e927b5c425529eab774093f6a8f9dba034e7220af62cd3bf6cefda1bef6f83.png" style="width: 1888px; height: 1417px;" /></a>
</div>
</div>
<p>Visualize genes for interpretation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize some gene expression on the above tsne:</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> 
           <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD34&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF8&#39;</span><span class="p">,</span> <span class="s1">&#39;ITGA2B&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79A&#39;</span><span class="p">],</span> <span class="n">ncols</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/f90854b110b7a14132ce77f55dc809d5e8bffd158d073b2704115745d9ece7a7.png"><img alt="_images/f90854b110b7a14132ce77f55dc809d5e8bffd158d073b2704115745d9ece7a7.png" src="_images/f90854b110b7a14132ce77f55dc809d5e8bffd158d073b2704115745d9ece7a7.png" style="width: 1211px; height: 721px;" /></a>
</div>
</div>
<section id="define-start-cell">
<h3>Define start cell<a class="headerlink" href="#define-start-cell" title="Link to this heading">#</a></h3>
<p>Palantir expects thatyou provide an example cell as a start point. The algorithm will use this cells as the initiating point of the trajectory but with each iteration, the start cell will be refined automatically. In other words, palantir is very robust to the choice of start cell, so it is OK that you only provide an approximate <code class="docutils literal notranslate"><span class="pre">start_cell</span></code>.</p>
<p>But how does one generally assign a start cell? This is a context dependent problem. In this data, the authors decided to find the cell with high expression of CD34 gene (the most primitive cell from Cluster 0) and define it as the start point. For this they selected a random cell from the highlighted region below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define a start point</span>
<span class="c1"># Please note that the start_cell should be the barcode of the cell estimated to be an ideal start point</span>
<span class="n">start_cell</span> <span class="o">=</span> <span class="s1">&#39;Run5_164698952452459&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">select_cells</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cells</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cells</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Cluster to select start cell&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">select_cells</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">start_cell</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cells</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cells</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Selected start cell&#39;</span><span class="p">);</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/ef7ef7bd6cf1cced5972577aaf1cf61791168b8635e676972fc1f309401d0630.png"><img alt="_images/ef7ef7bd6cf1cced5972577aaf1cf61791168b8635e676972fc1f309401d0630.png" src="_images/ef7ef7bd6cf1cced5972577aaf1cf61791168b8635e676972fc1f309401d0630.png" style="width: 1268px; height: 498px;" /></a>
</div>
</div>
</section>
<section id="define-terminal-states">
<h3>Define terminal states<a class="headerlink" href="#define-terminal-states" title="Link to this heading">#</a></h3>
<p>Palantir can automatically identify the terminal points, so you can begin your analysis by seeing what palantir identifies as terminal state(s). However, if you want to use biological intuition to identify possible terminal states, you can specify the cells which you think are likely terminal states.</p>
<p>If you want a data driven way to identify potential terminal states then you can study the cells with the minimum and maximum values of the diffusion components. Since Palantir builds on the diffusion components, you can study the components itself to identify potential end states. Now, we will color the tSNE by diffusion components, and highlight the cells with min and max values for each component.</p>
<p>Note: we only look at the first 8 diffusion component because the gap in eigenvalues is highest there, meaning that palantir will focus on these 8 components to derive a pseudotime.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors_multiscaled&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(5780, 7)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize the first 8 eigenvectors (Note: we ignore the 0th so 7 really)</span>
<span class="n">ncol</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">nrow</span> <span class="o">=</span> <span class="mi">2</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="n">ncol</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">nrow</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">8</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">nrow</span><span class="p">,</span> <span class="n">ncol</span><span class="p">,</span> <span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
                 <span class="n">c</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="n">j</span><span class="p">],</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
    
    <span class="p">[</span><span class="n">id_min</span><span class="p">,</span> <span class="n">id_max</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="n">j</span><span class="p">])[[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
    
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][</span><span class="n">id_min</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
               <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][</span><span class="n">id_min</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][</span><span class="n">id_max</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
               <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tsne&#39;</span><span class="p">][</span><span class="n">id_max</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
    
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Diff Map - &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">16</span><span class="p">)</span>
    
<span class="c1"># To save: </span>
<span class="c1"># fig.savefig(output_directory + &#39;umap_colored_by_diff_maps.png&#39;, bbox_inches = &#39;tight&#39;, dpi = 150)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/e9be17bbd190abc3e09f36fae945acd6e4b4fa25622e557cd7a13656c3385b56.png"><img alt="_images/e9be17bbd190abc3e09f36fae945acd6e4b4fa25622e557cd7a13656c3385b56.png" src="_images/e9be17bbd190abc3e09f36fae945acd6e4b4fa25622e557cd7a13656c3385b56.png" style="width: 2524px; height: 963px;" /></a>
</div>
</div>
<p>In the plot above the tsne representation of the data is colored by each diffusion component. The cells with minimum value for the diffusion component is highlighted in red, while the cells with the maximum value for the diffusion component is highlighted is black. For the plots above, we can speculate that the potential terminal states are as follows after taking care of duplicates:</p>
<ol class="arabic simple">
<li><p>min and max of diffusion component 1</p></li>
<li><p>max of diffusion component 2</p></li>
<li><p>max of diffusion component 3</p></li>
<li><p>max of diffusion component 5</p></li>
<li><p>max of diffusion component 6</p></li>
</ol>
<p>We are making this claim, also based on the gene expression and cell types in each of these terminal states. So, when you adapt this code for your own data, we suggest that you study your data based on gene expression and as above where we looked at the cells with minimum and maximum values for each diffusion component.</p>
<p>Let’s get the cell barcode for the cells corresponding to the minimum and maximum values of the above mentioned diffusion components.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">terminal_states</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># for component 1</span>
<span class="n">cell_comp1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">])[[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
<span class="n">terminal_states</span> <span class="o">=</span> <span class="n">terminal_states</span> <span class="o">+</span> <span class="nb">list</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">cell_comp1</span><span class="p">])</span>

<span class="c1"># for component 2</span>
<span class="n">cell_comp2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="mi">2</span><span class="p">])[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">terminal_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">cell_comp2</span><span class="p">])</span>


<span class="c1"># for component 3</span>
<span class="n">cell_comp3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="mi">3</span><span class="p">])[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">terminal_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">cell_comp3</span><span class="p">])</span>

<span class="c1"># for component 5</span>
<span class="n">cell_comp5</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="mi">5</span><span class="p">])[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">terminal_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">cell_comp5</span><span class="p">])</span>

<span class="c1"># for component 6</span>
<span class="n">cell_comp6</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;DM_EigenVectors&#39;</span><span class="p">][:,</span> <span class="mi">6</span><span class="p">])[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">terminal_states</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">cell_comp6</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">terminal_states</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Run4_239448276221238&#39;,
 &#39;Run4_236650105264556&#39;,
 &#39;Run5_239477254471070&#39;,
 &#39;Run4_130736318737710&#39;,
 &#39;Run5_191691434420467&#39;,
 &#39;Run4_200967489997038&#39;]
</pre></div>
</div>
</div>
</div>
<p>Visualize start and terminal states on tsne:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">select_cell_start</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">start_cell</span><span class="p">)</span>
<span class="n">select_cell_terminal</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">terminal_states</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cell_start</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
            <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cell_start</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;start&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cell_terminal</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
            <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">select_cell_terminal</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;terminal&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Selected start cell and terminal states&#39;</span><span class="p">);</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;matplotlib.legend.Legend at 0x2e874b400&gt;
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/1d066efd7f12e2766f84d953f9fec417e893b7b7fe69b6dbbb6e8ed31011d3bf.png"><img alt="_images/1d066efd7f12e2766f84d953f9fec417e893b7b7fe69b6dbbb6e8ed31011d3bf.png" src="_images/1d066efd7f12e2766f84d953f9fec417e893b7b7fe69b6dbbb6e8ed31011d3bf.png" style="width: 379px; height: 374px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize some gene expression on the above tsne:</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> 
           <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD34&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">,</span> <span class="s1">&#39;IRF8&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/4f4425c058728113cd0e14efd200cf717d367ebb761f7194eb690cfb3336279b.png"><img alt="_images/4f4425c058728113cd0e14efd200cf717d367ebb761f7194eb690cfb3336279b.png" src="_images/4f4425c058728113cd0e14efd200cf717d367ebb761f7194eb690cfb3336279b.png" style="width: 1599px; height: 373px;" /></a>
</div>
</div>
</section>
<section id="compute-pseudotime">
<h3>Compute pseudotime<a class="headerlink" href="#compute-pseudotime" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pr_res</span> <span class="o">=</span> <span class="n">palantir</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">run_palantir</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> 
                                    <span class="n">early_cell</span> <span class="o">=</span> <span class="n">start_cell</span><span class="p">,</span> 
                                    <span class="n">terminal_states</span><span class="o">=</span><span class="n">terminal_states</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sampling and flocking waypoints...
Time for determining waypoints: 0.005348682403564453 minutes
Determining pseudotime...
Shortest path distances using 30-nearest neighbor graph...
Time for shortest paths: 0.8863316337267558 minutes
Iteratively refining the pseudotime...
Correlation at iteration 1: 0.9996
Correlation at iteration 2: 1.0000
Entropy and branch probabilities...
Markov chain construction...
Computing fundamental matrix and absorption probabilities...
Project results to all cells...
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<p>Now to further process pr_res, we will follow: <a class="reference external" href="https://nbviewer.org/github/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb">https://nbviewer.org/github/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb</a></p>
<p>Palantir generates the following results</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">palantir_pseudotime</span></code>: Pseudo time ordering of each cell</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">palantir_fate_probabilities</span></code>: Matrix of cells X terminal states. Each entry represents the probability of the corresponding cell reaching the respective terminal state</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">palantir_entropy</span></code>: A quantiative measure of the differentiation potential of each cell computed as the entropy of the multinomial terminal state probabilities</p></li>
</ul>
</section>
<section id="get-the-pseudotime">
<h3>Get the pseudotime<a class="headerlink" href="#get-the-pseudotime" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;palantir_pseudotime&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>index
Run4_120703408880541    0.550738
Run4_120703409056541    0.074225
Run4_120703409580963    0.752128
Run4_120703423990708    0.722782
Run4_120703424252854    0.564369
                          ...   
Run5_241114589051630    0.857101
Run5_241114589051819    0.413827
Run5_241114589128940    0.472308
Run5_241114589357942    0.887548
Run5_241114589841822    0.470172
Name: palantir_pseudotime, Length: 5780, dtype: float64
</pre></div>
</div>
</div>
</div>
<p>In the output above, each row is a cell barcode and the number associated with each cell is it’s pseudotime alignment score (between 0 representing early and 1 representing differentiated).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the branch probabilities</span>
<span class="c1"># As shown below, palantir used the 6 terminally differentiated states</span>
<span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Run4_130736318737710</th>
      <th>Run4_200967489997038</th>
      <th>Run4_236650105264556</th>
      <th>Run4_239448276221238</th>
      <th>Run5_191691434420467</th>
      <th>Run5_239477254471070</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Run4_120703408880541</th>
      <td>0.166378</td>
      <td>0.033098</td>
      <td>0.692449</td>
      <td>0.000000</td>
      <td>0.035484</td>
      <td>0.068032</td>
    </tr>
    <tr>
      <th>Run4_120703409056541</th>
      <td>0.392431</td>
      <td>0.077788</td>
      <td>0.310796</td>
      <td>0.026885</td>
      <td>0.032207</td>
      <td>0.159892</td>
    </tr>
    <tr>
      <th>Run4_120703409580963</th>
      <td>0.938346</td>
      <td>0.011350</td>
      <td>0.023633</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.023331</td>
    </tr>
    <tr>
      <th>Run4_120703423990708</th>
      <td>0.874245</td>
      <td>0.023151</td>
      <td>0.048204</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.047587</td>
    </tr>
    <tr>
      <th>Run4_120703424252854</th>
      <td>0.406759</td>
      <td>0.118921</td>
      <td>0.201461</td>
      <td>0.000000</td>
      <td>0.021106</td>
      <td>0.244437</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Run5_241114589051630</th>
      <td>0.024373</td>
      <td>0.000000</td>
      <td>0.954936</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Run5_241114589051819</th>
      <td>0.436307</td>
      <td>0.086823</td>
      <td>0.261348</td>
      <td>0.000000</td>
      <td>0.027347</td>
      <td>0.178463</td>
    </tr>
    <tr>
      <th>Run5_241114589128940</th>
      <td>0.420652</td>
      <td>0.083913</td>
      <td>0.283884</td>
      <td>0.000000</td>
      <td>0.029924</td>
      <td>0.172480</td>
    </tr>
    <tr>
      <th>Run5_241114589357942</th>
      <td>0.014986</td>
      <td>0.000000</td>
      <td>0.972292</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Run5_241114589841822</th>
      <td>0.442553</td>
      <td>0.088615</td>
      <td>0.251219</td>
      <td>0.000000</td>
      <td>0.026313</td>
      <td>0.182146</td>
    </tr>
  </tbody>
</table>
<p>5780 rows × 6 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can rename the different branches:</span>
<span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Branch1&#39;</span><span class="p">,</span> <span class="s1">&#39;Branch2&#39;</span><span class="p">,</span> <span class="s1">&#39;Branch3&#39;</span><span class="p">,</span> <span class="s1">&#39;Branch4&#39;</span><span class="p">,</span> <span class="s1">&#39;Branch5&#39;</span><span class="p">,</span> <span class="s1">&#39;Branch6&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Get the entropy</span>
<span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;palantir_entropy&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>index
Run4_120703408880541    0.991592
Run4_120703409056541    1.429927
Run4_120703409580963    0.307683
Run4_120703423990708    0.533640
Run4_120703424252854    1.403655
                          ...   
Run5_241114589051630    0.238597
Run5_241114589051819    1.375757
Run5_241114589128940    1.380736
Run5_241114589357942    0.160424
Run5_241114589841822    1.371430
Name: palantir_entropy, Length: 5780, dtype: float64
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualize-the-pseudotime">
<h3>Visualize the pseudotime<a class="headerlink" href="#visualize-the-pseudotime" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;palantir_entropy&#39;</span><span class="p">,</span> <span class="s1">&#39;palantir_pseudotime&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/241c9d0ab6943943fa3750153398225035606066e28ff99bbf329949a3ead0a2.png"><img alt="_images/241c9d0ab6943943fa3750153398225035606066e28ff99bbf329949a3ead0a2.png" src="_images/241c9d0ab6943943fa3750153398225035606066e28ff99bbf329949a3ead0a2.png" style="width: 830px; height: 373px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize branch probability</span>
<span class="n">nrow</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">ncol</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="n">ncol</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="n">nrow</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pr_res</span><span class="o">.</span><span class="n">branch_probs</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">nrow</span><span class="p">,</span> <span class="n">ncol</span><span class="p">,</span> <span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> 
              <span class="n">c</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">][</span><span class="n">item</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s1">&#39;plasma&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Branch probability - &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">16</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/3db433cee5ba75c11caf35483c3c79004e9629a38d7f3ddb91a64ce9514ba49f.png"><img alt="_images/3db433cee5ba75c11caf35483c3c79004e9629a38d7f3ddb91a64ce9514ba49f.png" src="_images/3db433cee5ba75c11caf35483c3c79004e9629a38d7f3ddb91a64ce9514ba49f.png" style="width: 1866px; height: 966px;" /></a>
</div>
</div>
</section>
<section id="visualize-the-probability-for-cells-to-reach-certain-terminal-state">
<h3>Visualize the probability for cells to reach certain terminal state<a class="headerlink" href="#visualize-the-probability-for-cells-to-reach-certain-terminal-state" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">example_cells</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Run5_164698952452459&quot;</span><span class="p">,</span> <span class="s2">&quot;Run5_170327461775790&quot;</span><span class="p">,</span> <span class="s2">&quot;Run4_121896095574750&quot;</span><span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">cell_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">example_cells</span><span class="p">]</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> 
          <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">example_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> 
          <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;g&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">example_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_tsne&#39;</span><span class="p">][</span><span class="n">cell_id</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> 
          <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">example_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> 
       <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Branches&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Probability&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> 
       <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Branches&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Probability&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> 
       <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;palantir_fate_probabilities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">]])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Branches&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Probability&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">example_cells</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;Run4_121896095574750&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/697d321626cdf000e8e7ddc617f814673cd3b9fdc33746484721333d39bd55ff.png"><img alt="_images/697d321626cdf000e8e7ddc617f814673cd3b9fdc33746484721333d39bd55ff.png" src="_images/697d321626cdf000e8e7ddc617f814673cd3b9fdc33746484721333d39bd55ff.png" style="width: 1339px; height: 1005px;" /></a>
</div>
</div>
</section>
<section id="visualize-gene-expression-trends">
<h3>Visualize gene expression trends<a class="headerlink" href="#visualize-gene-expression-trends" title="Link to this heading">#</a></h3>
<p>Palantir allows us to study the gene expression trends along each of the branches or the pseudotime. This can help us identify regions along the pseudotime when gene expression change substantially, which can be informative about the biology of the system.</p>
<p>First, you can choose a branch to highlight:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">masks</span> <span class="o">=</span> <span class="n">palantir</span><span class="o">.</span><span class="n">presults</span><span class="o">.</span><span class="n">select_branch_cells</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;branch_masks&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Branch1</th>
      <th>Branch2</th>
      <th>Branch3</th>
      <th>Branch4</th>
      <th>Branch5</th>
      <th>Branch6</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Run4_120703408880541</th>
      <td>False</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run4_120703409056541</th>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run4_120703409580963</th>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run4_120703423990708</th>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run4_120703424252854</th>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Run5_241114589051630</th>
      <td>False</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run5_241114589051819</th>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Run5_241114589128940</th>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run5_241114589357942</th>
      <td>False</td>
      <td>False</td>
      <td>True</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
    <tr>
      <th>Run5_241114589841822</th>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>5780 rows × 6 columns</p>
</div></div></div>
</div>
<p>Visualizing the branch selection:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">palantir</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_branch_selection</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function plot_branch_selection in module palantir.plot:

plot_branch_selection(ad: anndata._core.anndata.AnnData, pseudo_time_key: str = &#39;palantir_pseudotime&#39;, fate_prob_key: str = &#39;palantir_fate_probabilities&#39;, masks_key: str = &#39;branch_masks&#39;, fates: Union[List[str], str, NoneType] = None, embedding_basis: str = &#39;X_umap&#39;, figsize: Tuple[float, float] = (15, 5), **kwargs)
    Plot cells along specific fates of pseudotime ordering and the UMAP embedding.
    
    Parameters
    ----------
    ad : sc.AnnData
        Annotated data matrix. The pseudotime and fate probabilities should be stored under the keys provided.
    pseudo_time_key : str, optional
        Key to access the pseudotime from obs of the AnnData object. Default is &#39;palantir_pseudotime&#39;.
    fate_prob_key : str, optional
        Key to access the fate probabilities from obsm of the AnnData object.
        Default is &#39;palantir_fate_probabilities&#39;.
    masks_key : str, optional
        Key to access the branch cell selection masks from obsm of the AnnData object.
        Default is &#39;branch_masks&#39;.
    fates : Union[List[str], str]
        The fates to be plotted. If not specified, all fates will be plotted.
    embedding_basis : str, optional
        Key to access the UMAP embedding from obsm of the AnnData object. Default is &#39;X_umap&#39;.
    figsize : Tuple[float, float], optional
        Width and height of each subplot in inches. The total height of the figure is determined by
        multiplying the height by the number of fates. Default is (15, 5).
    **kwargs
        Additional arguments passed to `matplotlib.pyplot.scatter`.
    
    Returns
    -------
    matplotlib.pyplot.Figure
        A matplotlib Figure object representing the plot of the branch selections.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palantir</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_branch_selection</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">embedding_basis</span> <span class="o">=</span> <span class="s1">&#39;tsne&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/b63cf419bdd29f9271eb4aeb10952ff871e6c49b977bc87d114a78c96ea57704.png"><img alt="_images/b63cf419bdd29f9271eb4aeb10952ff871e6c49b977bc87d114a78c96ea57704.png" src="_images/b63cf419bdd29f9271eb4aeb10952ff871e6c49b977bc87d114a78c96ea57704.png" style="width: 1478px; height: 2978px;" /></a>
</div>
</div>
<p>Now we can compute the gene expression trend along the trajectory.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_trends</span> <span class="o">=</span> <span class="n">palantir</span><span class="o">.</span><span class="n">presults</span><span class="o">.</span><span class="n">compute_gene_trends</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">)</span><span class="c1">#, expression_key=&quot;magic_imputed_data&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Branch1
[2024-11-07 22:58:32,311] [INFO    ] Using sparse Gaussian Process since n_landmarks (500) &lt; n_samples (1,444) and rank = 1.0.
[2024-11-07 22:58:32,312] [INFO    ] Using covariance function Matern52(ls=1.0).
Branch2
[2024-11-07 22:58:35,021] [INFO    ] Using sparse Gaussian Process since n_landmarks (500) &lt; n_samples (1,361) and rank = 1.0.
[2024-11-07 22:58:35,021] [INFO    ] Using covariance function Matern52(ls=1.0).
Branch3
[2024-11-07 22:58:36,951] [INFO    ] Using sparse Gaussian Process since n_landmarks (500) &lt; n_samples (1,240) and rank = 1.0.
[2024-11-07 22:58:36,953] [INFO    ] Using covariance function Matern52(ls=1.0).
Branch4
[2024-11-07 22:58:39,120] [INFO    ] Using sparse Gaussian Process since n_landmarks (500) &lt; n_samples (540) and rank = 1.0.
[2024-11-07 22:58:39,122] [INFO    ] Using covariance function Matern52(ls=1.0).
Branch5
[2024-11-07 22:58:40,712] [INFO    ] Using non-sparse Gaussian Process since n_landmarks (500) &gt;= n_samples (476) and rank = 1.0.
[2024-11-07 22:58:40,713] [INFO    ] Using covariance function Matern52(ls=1.0).
Branch6
[2024-11-07 22:58:42,948] [INFO    ] Using sparse Gaussian Process since n_landmarks (500) &lt; n_samples (1,370) and rank = 1.0.
[2024-11-07 22:58:42,949] [INFO    ] Using covariance function Matern52(ls=1.0).
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CD34&quot;</span><span class="p">,</span> <span class="s2">&quot;MPO&quot;</span><span class="p">,</span> <span class="s2">&quot;GATA1&quot;</span><span class="p">,</span> <span class="s2">&quot;IRF8&quot;</span><span class="p">]</span>
<span class="n">palantir</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_gene_trends</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/38ba2b6b4eb5e08c76b8957443dd8530726de0c879aadea11896701e79c1e23e.png"><img alt="_images/38ba2b6b4eb5e08c76b8957443dd8530726de0c879aadea11896701e79c1e23e.png" src="_images/38ba2b6b4eb5e08c76b8957443dd8530726de0c879aadea11896701e79c1e23e.png" style="width: 613px; height: 982px;" /></a>
</div>
</div>
<p>As heatmaps:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palantir</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_gene_trend_heatmaps</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/8c88a2bf8b23e6760c40f4b71a1510188189a4f2a4be075c1ba482a52a1b6218.png"><img alt="_images/8c88a2bf8b23e6760c40f4b71a1510188189a4f2a4be075c1ba482a52a1b6218.png" src="_images/8c88a2bf8b23e6760c40f4b71a1510188189a4f2a4be075c1ba482a52a1b6218.png" style="width: 594px; height: 1313px;" /></a>
</div>
</div>
</section>
<section id="clustering-gene-expression-trends">
<h3>Clustering gene expression trends<a class="headerlink" href="#clustering-gene-expression-trends" title="Link to this heading">#</a></h3>
<p>It may also be informative to group genes together based on how they trend along the pseudotime. For exposition, we will randomly sample 1000 genes from the data and cluster them based on how they trend along Branch1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">random</span>
<span class="n">sample_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">var_names</span><span class="p">),</span> <span class="mi">1000</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">palantir</span><span class="o">.</span><span class="n">presults</span><span class="o">.</span><span class="n">cluster_gene_trends</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function cluster_gene_trends in module palantir.presults:

cluster_gene_trends(data: Union[anndata._core.anndata.AnnData, pandas.core.frame.DataFrame], branch_name: str, genes: Optional[List[str]] = None, gene_trend_key: Optional[str] = &#39;gene_trends&#39;, n_neighbors: int = 150, **kwargs) -&gt; pandas.core.series.Series
    Cluster gene trends using the Leiden algorithm.
    
    This function applies the Leiden clustering algorithm to gene expression trends
    along the pseudotemporal trajectory. If the input is an AnnData object, it uses
    the gene trends stored in the `varm` attribute accessed using the `gene_trend_key`.
    If the input is a DataFrame, it directly uses the input data for clustering.
    
    Parameters
    ----------
    data : Union[sc.AnnData, pd.DataFrame]
        AnnData object or a DataFrame of gene expression trends.
    branch_name : str
        Name of the branch for which the gene trends are to be clustered.
    genes : list of str, optional
        List of genes to be considered for clustering. If None, all genes are considered. Default is None.
    gene_trend_key : str, optional
        Key to access gene trends in the AnnData object&#39;s varm. Default is &#39;palantir_gene_trends&#39;.
    n_neighbors : int, optional
        The number of nearest neighbors to use for the k-NN graph construction. Default is 150.
    **kwargs
        Additional keyword arguments passed to `scanpy.tl.leiden`.
    
    Returns
    -------
    pd.Series
        A pandas series with the cluser lables for all passed genes.
    
    Raises
    ------
    KeyError
        If `gene_trend_key` is None when `data` is an AnnData object.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">communities</span> <span class="o">=</span> <span class="n">palantir</span><span class="o">.</span><span class="n">presults</span><span class="o">.</span><span class="n">cluster_gene_trends</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> 
                                                    <span class="n">branch_name</span> <span class="o">=</span> <span class="s2">&quot;Branch1&quot;</span><span class="p">,</span> 
                                                    <span class="n">genes</span> <span class="o">=</span> <span class="n">sample_genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palantir</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">plot_gene_trend_clusters</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="s2">&quot;Branch1&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/a04ea7179f6706d17d52752db6f823c11b1b4275472e42540085995322f3d181.png"><img alt="_images/a04ea7179f6706d17d52752db6f823c11b1b4275472e42540085995322f3d181.png" src="_images/a04ea7179f6706d17d52752db6f823c11b1b4275472e42540085995322f3d181.png" style="width: 1339px; height: 800px;" /></a>
</div>
</div>
</section>
</section>
<section id="paga">
<h2>PAGA<a class="headerlink" href="#paga" title="Link to this heading">#</a></h2>
<p>PAGA (PArtition-based Graph Abstraction) is another method that can be used to study connectivities in single-cell data. Often, studying cell differentiation process using pseudotime may not be optimal due to issues resulting from inadequate sampling of cells. In such cases, the notion of connecting clusters of cells using graph-based metrics (to avoid errors that Euclidean distance might cause) has been proposed as an alternative, and PAGA is one such method. PAGA attempts to preserve both continuous and disconnected structure in the data.</p>
<p>Ref: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6425583/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6425583/</a></p>
<p>PAGA relies on both clustering and nearest neighbor graphs (which can be constructed on PCA or on diffusion components). We will provide code to do both.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    var: &#39;gene_trends_clusters&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">tsne</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;clusters&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/Users/sara.jimenez/miniconda3/envs/scvi-env/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/78440b84997d129ce17460b90cae0153eb3c1ae034ff842df08afc89251edffa.png"><img alt="_images/78440b84997d129ce17460b90cae0153eb3c1ae034ff842df08afc89251edffa.png" src="_images/78440b84997d129ce17460b90cae0153eb3c1ae034ff842df08afc89251edffa.png" style="width: 404px; height: 369px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We are computing 300 principal components as that is what the authors used. </span>
<span class="c1"># The rationale was that they wanted to consider enough PCs to explain 75% of the variance in the data</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_comps</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute nearest neighbors </span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span> <span class="o">=</span> <span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="paga-with-distances-computed-on-pca">
<h3>PAGA with distances computed on PCA<a class="headerlink" href="#paga-with-distances-computed-on-pca" title="Link to this heading">#</a></h3>
<p>Since we have already computed k-nearest neighbor distance graph on the PCA, we can simply invoke that result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s1">&#39;clusters&#39;</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    var: &#39;gene_trends_clusters&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;, &#39;paga&#39;, &#39;clusters_sizes&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;paga&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;connectivities&#39;: &lt;10x10 sparse matrix of type &#39;&lt;class &#39;numpy.float64&#39;&gt;&#39;
 	with 74 stored elements in Compressed Sparse Row format&gt;,
 &#39;connectivities_tree&#39;: &lt;10x10 sparse matrix of type &#39;&lt;class &#39;numpy.float64&#39;&gt;&#39;
 	with 9 stored elements in Compressed Sparse Row format&gt;,
 &#39;groups&#39;: &#39;clusters&#39;}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">],</span> <span class="n">node_size_scale</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/466677740458c89dda010cfe97418b6ad7ab263428e3dd13d44dd58e25400a06.png"><img alt="_images/466677740458c89dda010cfe97418b6ad7ab263428e3dd13d44dd58e25400a06.png" src="_images/466677740458c89dda010cfe97418b6ad7ab263428e3dd13d44dd58e25400a06.png" style="width: 331px; height: 352px;" /></a>
</div>
</div>
<p>Interpreting the plot above:</p>
<ol class="arabic simple">
<li><p>Each circle indicates a cluster and the size of the circle is proportional to size of the cluster.</p></li>
<li><p>Two circles that are phenotypically close to each other - as quantified by the distances computed on the PCA space - are connected to each other.</p></li>
<li><p>The width of the connecting edges is proportional to the similarity between the clusters.</p></li>
</ol>
<p>We can also highlight the expression of select genes onto the PAGA plot as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD34&#39;</span><span class="p">,</span> <span class="s1">&#39;MPO&#39;</span><span class="p">,</span> <span class="s1">&#39;GATA1&#39;</span><span class="p">],</span> <span class="n">node_size_scale</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/0485292fff5cb7ca4b68477fbc19c3e165f2c1494107cf364472338ece5cc996.png"><img alt="_images/0485292fff5cb7ca4b68477fbc19c3e165f2c1494107cf364472338ece5cc996.png" src="_images/0485292fff5cb7ca4b68477fbc19c3e165f2c1494107cf364472338ece5cc996.png" style="width: 1154px; height: 374px;" /></a>
</div>
</div>
</section>
<section id="paga-on-diffusion-components">
<h3>PAGA on diffusion components<a class="headerlink" href="#paga-on-diffusion-components" title="Link to this heading">#</a></h3>
<p>As we discussed above, PAGA can also compute connectivites between clusters based on distances in the diffusion space. For the example, below we will consider distances in the diffusion space defined by the first seven eigenvectors (ignoring the 0th eigenvector – automatically ignored by PAGA).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    var: &#39;gene_trends_clusters&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;, &#39;paga&#39;, &#39;clusters_sizes&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">diffmap</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])),</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Index&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;EigenValue&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="s1">&#39;on&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/1bb6de518dac06beb20e4f02daa466602cc4fd8d3faca8437a6a3bcef4a0ec09.png"><img alt="_images/1bb6de518dac06beb20e4f02daa466602cc4fd8d3faca8437a6a3bcef4a0ec09.png" src="_images/1bb6de518dac06beb20e4f02daa466602cc4fd8d3faca8437a6a3bcef4a0ec09.png" style="width: 398px; height: 375px;" /></a>
</div>
</div>
<p>Based on the above plot, we will consider only the top 7 components:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_diffmap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
<span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;diffmap_evals&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(5780, 7)
</pre></div>
</div>
</div>
</div>
<p>Compute neighbors in diffusion components space:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">,</span> <span class="n">key_added</span> <span class="o">=</span> <span class="s1">&#39;diff_neighbors&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Re-do PAGA on diffusion map space:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s1">&#39;clusters&#39;</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;diff_neighbors&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    var: &#39;gene_trends_clusters&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;, &#39;paga&#39;, &#39;clusters_sizes&#39;, &#39;diff_neighbors&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;, &#39;diffmap_evecs&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;, &#39;diff_neighbors_distances&#39;, &#39;diff_neighbors_connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">paga</span><span class="p">(</span><span class="n">adata_bm</span><span class="p">,</span> <span class="n">node_size_scale</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/772957f77d4d24a4f3d7b13442dd3787480aa7e03c1e7640d68ec0239553d2de.png"><img alt="_images/772957f77d4d24a4f3d7b13442dd3787480aa7e03c1e7640d68ec0239553d2de.png" src="_images/772957f77d4d24a4f3d7b13442dd3787480aa7e03c1e7640d68ec0239553d2de.png" style="width: 331px; height: 352px;" /></a>
</div>
</div>
<p>As you can see the two results of PAGA look visually quite similar (except that cluster 9 was more connected to other clusters when computing distances in the PCA space vs when computing in the diffusion space). It will be up to you to evaluate such cases using what is known about the cell types. In this case, Cluster 9 is ITGA2B+ (Megakaryocytes) and it is more phenotypically closer to cluster 2 and 8 as opposed to clusters 1 and 4. Therefore, diffusion distance based PAGA would be more faithful representation of the data here.</p>
</section>
</section>
<section id="using-results-to-quantify-distance-between-clusters">
<h2>Using results to quantify distance between clusters<a class="headerlink" href="#using-results-to-quantify-distance-between-clusters" title="Link to this heading">#</a></h2>
<p>You can extract the similarity between clusters using diffusion maps space as follows. We see 0 when two clusters are not connected to each other in PAGA.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;paga&#39;</span><span class="p">][</span><span class="s1">&#39;connectivities&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">(),</span>
             <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cluster-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">])],</span> 
             <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Cluster-&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">])])</span>

<span class="n">df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster-0</th>
      <th>Cluster-1</th>
      <th>Cluster-2</th>
      <th>Cluster-3</th>
      <th>Cluster-4</th>
      <th>Cluster-5</th>
      <th>Cluster-6</th>
      <th>Cluster-7</th>
      <th>Cluster-8</th>
      <th>Cluster-9</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Cluster-0</th>
      <td>0.000000</td>
      <td>0.335746</td>
      <td>0.042759</td>
      <td>0.008247</td>
      <td>0.004014</td>
      <td>0.008583</td>
      <td>0.000415</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.001461</td>
    </tr>
    <tr>
      <th>Cluster-1</th>
      <td>0.335746</td>
      <td>0.000000</td>
      <td>0.014329</td>
      <td>0.140419</td>
      <td>0.569995</td>
      <td>0.011620</td>
      <td>0.204404</td>
      <td>0.125988</td>
      <td>0.000000</td>
      <td>0.003320</td>
    </tr>
    <tr>
      <th>Cluster-2</th>
      <td>0.042759</td>
      <td>0.014329</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.083675</td>
      <td>0.000000</td>
      <td>0.000626</td>
      <td>0.000000</td>
      <td>0.564495</td>
      <td>0.917967</td>
    </tr>
    <tr>
      <th>Cluster-3</th>
      <td>0.008247</td>
      <td>0.140419</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.193001</td>
      <td>0.006204</td>
      <td>0.940253</td>
      <td>0.005905</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cluster-4</th>
      <td>0.004014</td>
      <td>0.569995</td>
      <td>0.083675</td>
      <td>0.193001</td>
      <td>0.000000</td>
      <td>0.001533</td>
      <td>0.678286</td>
      <td>0.178738</td>
      <td>0.000000</td>
      <td>0.007358</td>
    </tr>
    <tr>
      <th>Cluster-5</th>
      <td>0.008583</td>
      <td>0.011620</td>
      <td>0.000000</td>
      <td>0.006204</td>
      <td>0.001533</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cluster-6</th>
      <td>0.000415</td>
      <td>0.204404</td>
      <td>0.000626</td>
      <td>0.940253</td>
      <td>0.678286</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.411622</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cluster-7</th>
      <td>0.000000</td>
      <td>0.125988</td>
      <td>0.000000</td>
      <td>0.005905</td>
      <td>0.178738</td>
      <td>0.000000</td>
      <td>0.411622</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cluster-8</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.564495</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>Cluster-9</th>
      <td>0.001461</td>
      <td>0.003320</td>
      <td>0.917967</td>
      <td>0.000000</td>
      <td>0.007358</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Note: We can always repurpose this code to define distance or similarity between any two groups of cells by just replacing <code class="docutils literal notranslate"><span class="pre">clusters</span></code> column in obs with whatever grouping of cells we want.</p>
<p>Alternatively, for data where cells are along a continuum, we can use the diffusion components to compute phenotypic distances between cells or clusters. As we discussed, Euclidean distance in the diffusion component space is a good approximation for phenotypic distance between cells. Of course, we need to be careful with only the meaningful diffusion components (e.g. in the above case we will use diffusion components 1-7).</p>
<p>In the following code, we will compute distance (A) between each pair of cells within cluster 0 (B) between cells in cluster 0 and cluster 1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5780 × 14651
    obs: &#39;clusters&#39;, &#39;palantir_pseudotime&#39;, &#39;palantir_diff_potential&#39;, &#39;palantir_entropy&#39;
    var: &#39;gene_trends_clusters&#39;
    uns: &#39;cluster_colors&#39;, &#39;ct_colors&#39;, &#39;palantir_branch_probs_cell_types&#39;, &#39;clusters_colors&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;diffmap_evals&#39;, &#39;DM_EigenValues&#39;, &#39;palantir_waypoints&#39;, &#39;paga&#39;, &#39;clusters_sizes&#39;, &#39;diff_neighbors&#39;
    obsm: &#39;tsne&#39;, &#39;MAGIC_imputed_data&#39;, &#39;palantir_branch_probs&#39;, &#39;X_tsne&#39;, &#39;X_pca&#39;, &#39;X_diffmap&#39;, &#39;DM_EigenVectors&#39;, &#39;DM_EigenVectors_multiscaled&#39;, &#39;palantir_fate_probabilities&#39;, &#39;branch_masks&#39;, &#39;diffmap_evecs&#39;
    varm: &#39;PCs&#39;, &#39;gene_trends_Branch1&#39;, &#39;gene_trends_Branch2&#39;, &#39;gene_trends_Branch3&#39;, &#39;gene_trends_Branch4&#39;, &#39;gene_trends_Branch5&#39;, &#39;gene_trends_Branch6&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;DM_Kernel&#39;, &#39;DM_Similarity&#39;, &#39;diff_neighbors_distances&#39;, &#39;diff_neighbors_connectivities&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(5780, 7)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cells_C0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">cells_C1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;clusters&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cells_C0</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells_C1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1118 984
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">pairwise_distances</span>
<span class="n">distance_pairwise_C0</span> <span class="o">=</span> <span class="n">pairwise_distances</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">][</span><span class="n">cells_C0</span><span class="p">,</span> <span class="p">:])</span>
<span class="n">distance_pairwise_C0_C1</span> <span class="o">=</span> <span class="n">pairwise_distances</span><span class="p">(</span><span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">][</span><span class="n">cells_C0</span><span class="p">,</span> <span class="p">:],</span> 
                                             <span class="n">adata_bm</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;diffmap_evecs&#39;</span><span class="p">][</span><span class="n">cells_C1</span><span class="p">,</span> <span class="p">:])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">distance_pairwise_C0</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">distance_pairwise_C0_C1</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(1118, 1118) (1118, 984)
</pre></div>
</div>
</div>
</div>
<p>As we see, <code class="docutils literal notranslate"><span class="pre">distance_pairwise_C0</span></code> holds the pairwise distance between cells in cluster 0; <code class="docutils literal notranslate"><span class="pre">distance_pairwise_C0_C1</span></code> holds the pairwise distance between cells in cluster 0 and cluster 1.</p>
<p>We can visualize the distances as a boxplot for comparison. Also, note that <code class="docutils literal notranslate"><span class="pre">distance_pairwise_C0</span></code> is a symmetric matrix as distance from cell1 in cluster 0 to cell2 in cluster 0 is the same as distance from cell2 in cluster0 to cell1 in cluster 0. So we will only look at the upper traingle of the <code class="docutils literal notranslate"><span class="pre">distance_pairwise_C0</span></code> matrix.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">within_C0</span> <span class="o">=</span> <span class="n">distance_pairwise_C0</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">triu_indices</span><span class="p">(</span><span class="n">distance_pairwise_C0</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)]</span>
<span class="n">between_C0_C1</span> <span class="o">=</span> <span class="n">distance_pairwise_C0_C1</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">boxplot</span><span class="p">([</span><span class="n">within_C0</span><span class="p">,</span> <span class="n">between_C0_C1</span><span class="p">]);</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;Distance inside C0&#39;</span><span class="p">,</span> <span class="s1">&#39;Distance between C0 and C1&#39;</span><span class="p">]);</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Distance in diffusion space&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;Distance in diffusion space&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/a89a7408d5603e7c635315a097408e7a29a5687e5173d12de3b9fed04097df51.png"><img alt="_images/a89a7408d5603e7c635315a097408e7a29a5687e5173d12de3b9fed04097df51.png" src="_images/a89a7408d5603e7c635315a097408e7a29a5687e5173d12de3b9fed04097df51.png" style="width: 451px; height: 351px;" /></a>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="session-2.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Dimensionality reduction, clustering and annotation</p>
      </div>
    </a>
    <a class="right-next"
       href="session-3.2.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multiple samples</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Trajectory analysis</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-packages">Load packages</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-processed-data">Load processed data</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#trajectory-inference">Trajectory Inference</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#diffusion-maps">Diffusion Maps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-scanpy-inbuilt-diffusion-maps">Method 1: Scanpy inbuilt diffusion maps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-data-2-cells-along-a-continuum">Example data 2: Cells along a continuum</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#nearest-neighbor-and-diffusion-maps">Nearest neighbor and diffusion maps</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#palantir">Palantir</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-start-cell">Define start cell</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#define-terminal-states">Define terminal states</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-pseudotime">Compute pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#get-the-pseudotime">Get the pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-pseudotime">Visualize the pseudotime</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-probability-for-cells-to-reach-certain-terminal-state">Visualize the probability for cells to reach certain terminal state</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-gene-expression-trends">Visualize gene expression trends</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#clustering-gene-expression-trends">Clustering gene expression trends</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#paga">PAGA</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paga-with-distances-computed-on-pca">PAGA with distances computed on PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paga-on-diffusion-components">PAGA on diffusion components</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#using-results-to-quantify-distance-between-clusters">Using results to quantify distance between clusters</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sara Jimenez
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>