
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Introduction &#8212; GSCN workshop 2024</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!-- 
    this give us a css class that will be invisible only if js is disabled 
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=26a4bc78f4c0ddb94549" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=26a4bc78f4c0ddb94549"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549" />

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'session-3.2';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">GSCN workshop 2024</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    GSCN workshop 2024
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Session 1</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="session-1.html">Introduction</a></li>








</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/sarajimenez/GSCN_workshop_2024" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/session-3.2.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Introduction</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Introduction</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-packages">Load packages</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-multiple-samples-into-scanpy">Loading multiple samples into Scanpy</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-1">batch-1</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-2">batch-2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concatenate">Concatenate</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-qc-and-normalization">Basic QC and normalization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#highly-variable-genes-pca-nearest-neighbor-graph-umap-fdl-phenograph">Highly variable genes, PCA, Nearest Neighbor Graph, UMAP, FDL, Phenograph</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-results">Visualize the results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-some-genes">Visualize some genes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#highlight-b-cells">Highlight B cells</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction">Batch Effect Correction</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-scanorama">Batch effect correction - Scanorama</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nearest-neighbors-umap-fdl">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize">Visualize</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#true-celltypes">True celltypes</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-harmony">Batch effect correction - Harmony</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-fastmnn">Batch effect correction - FastMNN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-with-hvgs">Implementation with HVGs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#process-the-corrected-output">Process the corrected output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-data">Visualize the data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-with-all-genes">Implementation with all genes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Process the corrected output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Visualize the data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-seurat-integration">Batch effect correction - Seurat Integration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation">Implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#integrate">Integrate</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#study-the-output">Study the output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-to-adata">Add to adata</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-nearest-neighbors-umap-fdl">PCA, Nearest Neighbors, UMAP, FDL</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Visualize</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#ideas-to-quantifying-effectiveness-of-batch-effect-correction">Ideas to quantifying effectiveness of batch effect correction</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#before-correction">Before correction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#after-correction">After correction</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h1>
<p>In this notebook, we will work with multiple samples, discuss batch effect issues, use MiloR to measure differential cell abundance between conditions and discuss applicability of SEACells for data aggregation.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="load-packages">
<h1>Load packages<a class="headerlink" href="#load-packages" title="Link to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="loading-multiple-samples-into-scanpy">
<h1>Loading multiple samples into Scanpy<a class="headerlink" href="#loading-multiple-samples-into-scanpy" title="Link to this heading">#</a></h1>
<p>All of the computation we have done so far has been on a single sample. However, most real world data are likely not going to be a single sample and often require us to merge multiple data. We will use Scanpy to concatenate multiple data into one unified anndata object and discuss the issue of technical effect and its correction.</p>
<p>For the purposes of illustration we will use the PBMC data (dataset 5) used in this publication: <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9</a>. The authors provide the data in an easily usable format (<a class="github reference external" href="https://github.com/JinmiaoChenLab/Batch-effect-removal-benchmarking">JinmiaoChenLab/Batch-effect-removal-benchmarking</a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_path</span> <span class="o">=</span> <span class="s1">&#39;/data/peer/sharmar1/msk_workshop/workshop_2024_spring/session-4/&#39;</span>
</pre></div>
</div>
</div>
</div>
<section id="batch-1">
<h2>batch-1<a class="headerlink" href="#batch-1" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_batch1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;b1_exprs.txt&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">celltype_batch1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;b1_celltype.txt&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_batch1</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>data_3p-AAACCTGAGCATCATC-0</th>
      <th>data_3p-AAACCTGAGCTAGTGG-0</th>
      <th>data_3p-AAACCTGCACATTAGC-0</th>
      <th>data_3p-AAACCTGCACTGTTAG-0</th>
      <th>data_3p-AAACCTGCATAGTAAG-0</th>
      <th>data_3p-AAACCTGCATGAACCT-0</th>
      <th>data_3p-AAACCTGGTAAGAGGA-0</th>
      <th>data_3p-AAACCTGGTAGAAGGA-0</th>
      <th>data_3p-AAACCTGGTCCAGTGC-0</th>
      <th>data_3p-AAACCTGGTGTCTGAT-0</th>
      <th>...</th>
      <th>data_3p-TTTGTCACAGGGATTG-0</th>
      <th>data_3p-TTTGTCAGTAGCAAAT-0</th>
      <th>data_3p-TTTGTCAGTCAGATAA-0</th>
      <th>data_3p-TTTGTCAGTCGCGTGT-0</th>
      <th>data_3p-TTTGTCAGTTACCGAT-0</th>
      <th>data_3p-TTTGTCATCATGTCCC-0</th>
      <th>data_3p-TTTGTCATCCGATATG-0</th>
      <th>data_3p-TTTGTCATCGTCTGAA-0</th>
      <th>data_3p-TTTGTCATCTCGAGTA-0</th>
      <th>data_3p-TTTGTCATCTGCTTGC-0</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>RP11-34P13.3</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>FAM138A</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>OR4F5</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>RP11-34P13.7</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>RP11-34P13.8</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 8098 columns</p>
</div></div></div>
</div>
<p>We see above that the rows are genes and the columns are cells. We will have to keep this in mind when we create an anndata object out of the data.</p>
<p>Let’s create an anndata object for batch1:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Notes: </span>
<span class="c1"># We transpose the data to ensure rows are cells and columns are genes</span>
<span class="c1"># We consider columns of the dataframe above as the cell names</span>
<span class="c1"># We consider rows of the dataframe above as the gene names</span>
<span class="n">adata1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">data_batch1</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> 
                    <span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">data_batch1</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> 
                    <span class="n">var</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">data_batch1</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 8098 × 33694
</pre></div>
</div>
</div>
</div>
<p>Add the author provided celltype information:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_batch1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>n_counts</th>
      <th>n_genes</th>
      <th>batch</th>
      <th>louvain</th>
      <th>anno</th>
      <th>Method</th>
      <th>CellType</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>data_3p-AAACCTGAGCATCATC-0</th>
      <td>data_3p</td>
      <td>2394</td>
      <td>871</td>
      <td>0</td>
      <td>9</td>
      <td>B cell</td>
      <td>10X_3prime</td>
      <td>B cell</td>
    </tr>
    <tr>
      <th>data_3p-AAACCTGAGCTAGTGG-0</th>
      <td>data_3p</td>
      <td>4520</td>
      <td>1316</td>
      <td>0</td>
      <td>5</td>
      <td>CD4 T cell</td>
      <td>10X_3prime</td>
      <td>CD4 T cell</td>
    </tr>
    <tr>
      <th>data_3p-AAACCTGCACATTAGC-0</th>
      <td>data_3p</td>
      <td>2788</td>
      <td>898</td>
      <td>0</td>
      <td>1</td>
      <td>CD4 T cell</td>
      <td>10X_3prime</td>
      <td>CD4 T cell</td>
    </tr>
    <tr>
      <th>data_3p-AAACCTGCACTGTTAG-0</th>
      <td>data_3p</td>
      <td>4667</td>
      <td>1526</td>
      <td>0</td>
      <td>0</td>
      <td>Monocyte_CD14</td>
      <td>10X_3prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>data_3p-AAACCTGCATAGTAAG-0</th>
      <td>data_3p</td>
      <td>4440</td>
      <td>1495</td>
      <td>0</td>
      <td>0</td>
      <td>Monocyte_CD14</td>
      <td>10X_3prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>data_3p-TTTGTCATCATGTCCC-0</th>
      <td>data_3p</td>
      <td>3141</td>
      <td>1176</td>
      <td>0</td>
      <td>4</td>
      <td>CD8 T cell</td>
      <td>10X_3prime</td>
      <td>CD8 T cell</td>
    </tr>
    <tr>
      <th>data_3p-TTTGTCATCCGATATG-0</th>
      <td>data_3p</td>
      <td>5401</td>
      <td>1379</td>
      <td>0</td>
      <td>4</td>
      <td>CD8 T cell</td>
      <td>10X_3prime</td>
      <td>CD8 T cell</td>
    </tr>
    <tr>
      <th>data_3p-TTTGTCATCGTCTGAA-0</th>
      <td>data_3p</td>
      <td>6081</td>
      <td>1802</td>
      <td>0</td>
      <td>0</td>
      <td>Monocyte_CD14</td>
      <td>10X_3prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>data_3p-TTTGTCATCTCGAGTA-0</th>
      <td>data_3p</td>
      <td>3970</td>
      <td>1317</td>
      <td>0</td>
      <td>7</td>
      <td>CD8 T cell</td>
      <td>10X_3prime</td>
      <td>CD8 T cell</td>
    </tr>
    <tr>
      <th>data_3p-TTTGTCATCTGCTTGC-0</th>
      <td>data_3p</td>
      <td>4027</td>
      <td>1259</td>
      <td>0</td>
      <td>4</td>
      <td>CD8 T cell</td>
      <td>10X_3prime</td>
      <td>CD8 T cell</td>
    </tr>
  </tbody>
</table>
<p>8098 rows × 8 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">celltype_batch1</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">adata1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="n">celltype_batch1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata1</span><span class="o">.</span><span class="n">obs_names</span><span class="p">][</span><span class="n">item</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 8098 × 33694
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="batch-2">
<h2>batch-2<a class="headerlink" href="#batch-2" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_batch2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;b2_exprs.txt&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">celltype_batch2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_path</span> <span class="o">+</span> <span class="s1">&#39;b2_celltype.txt&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_batch2</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>data_5p-AAACCTGAGCGATAGC-1</th>
      <th>data_5p-AAACCTGAGCTAAACA-1</th>
      <th>data_5p-AAACCTGAGGGAGTAA-1</th>
      <th>data_5p-AAACCTGAGTCTTGCA-1</th>
      <th>data_5p-AAACCTGAGTTCGATC-1</th>
      <th>data_5p-AAACCTGCACACTGCG-1</th>
      <th>data_5p-AAACCTGCACGGTGTC-1</th>
      <th>data_5p-AAACCTGCAGATGGGT-1</th>
      <th>data_5p-AAACCTGCAGGTGGAT-1</th>
      <th>data_5p-AAACCTGGTAAGCACG-1</th>
      <th>...</th>
      <th>data_5p-TTTGTCACAGCTGGCT-1</th>
      <th>data_5p-TTTGTCACAGGTGGAT-1</th>
      <th>data_5p-TTTGTCAGTCCGAAGA-1</th>
      <th>data_5p-TTTGTCAGTTGATTGC-1</th>
      <th>data_5p-TTTGTCATCACAAACC-1</th>
      <th>data_5p-TTTGTCATCCACGTTC-1</th>
      <th>data_5p-TTTGTCATCGCGTAGC-1</th>
      <th>data_5p-TTTGTCATCTTAACCT-1</th>
      <th>data_5p-TTTGTCATCTTACCGC-1</th>
      <th>data_5p-TTTGTCATCTTGTTTG-1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>RP11-34P13.3</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>FAM138A</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>OR4F5</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>RP11-34P13.7</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>RP11-34P13.8</th>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>...</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 7378 columns</p>
</div></div></div>
</div>
<p>Similar to batch-1, we see above that the rows are genes and the columns are cells. We will have to keep this in mind when we create an anndata object out of the data.</p>
<p>Let’s create an anndata object for batch-2:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Notes: </span>
<span class="c1"># We transpose the data to ensure rows are cells and columns are genes</span>
<span class="c1"># We consider columns of the dataframe above as the cell names</span>
<span class="c1"># We consider rows of the dataframe above as the gene names</span>
<span class="n">adata2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">data_batch2</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> 
                    <span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">data_batch2</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> 
                    <span class="n">var</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">data_batch2</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 7378 × 33694
</pre></div>
</div>
</div>
</div>
<p>Add the author provided celltype information:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_batch2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sample</th>
      <th>n_counts</th>
      <th>n_genes</th>
      <th>batch</th>
      <th>louvain</th>
      <th>anno</th>
      <th>Method</th>
      <th>CellType</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>data_5p-AAACCTGAGCGATAGC-1</th>
      <td>data_5p</td>
      <td>2712</td>
      <td>1318</td>
      <td>1</td>
      <td>18</td>
      <td>NK cell</td>
      <td>10X_5prime</td>
      <td>NK cell</td>
    </tr>
    <tr>
      <th>data_5p-AAACCTGAGCTAAACA-1</th>
      <td>data_5p</td>
      <td>6561</td>
      <td>2164</td>
      <td>1</td>
      <td>3</td>
      <td>Monocyte_CD14</td>
      <td>10X_5prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>data_5p-AAACCTGAGGGAGTAA-1</th>
      <td>data_5p</td>
      <td>6322</td>
      <td>2112</td>
      <td>1</td>
      <td>8</td>
      <td>Monocyte_CD14</td>
      <td>10X_5prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>data_5p-AAACCTGAGTCTTGCA-1</th>
      <td>data_5p</td>
      <td>4528</td>
      <td>1526</td>
      <td>1</td>
      <td>16</td>
      <td>CD8 T cell</td>
      <td>10X_5prime</td>
      <td>CD8 T cell</td>
    </tr>
    <tr>
      <th>data_5p-AAACCTGAGTTCGATC-1</th>
      <td>data_5p</td>
      <td>3426</td>
      <td>1332</td>
      <td>1</td>
      <td>3</td>
      <td>Monocyte_CD14</td>
      <td>10X_5prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>data_5p-TTTGTCATCCACGTTC-1</th>
      <td>data_5p</td>
      <td>6547</td>
      <td>2044</td>
      <td>1</td>
      <td>3</td>
      <td>Monocyte_CD14</td>
      <td>10X_5prime</td>
      <td>Monocyte_CD14</td>
    </tr>
    <tr>
      <th>data_5p-TTTGTCATCGCGTAGC-1</th>
      <td>data_5p</td>
      <td>3615</td>
      <td>1397</td>
      <td>1</td>
      <td>10</td>
      <td>B cell</td>
      <td>10X_5prime</td>
      <td>B cell</td>
    </tr>
    <tr>
      <th>data_5p-TTTGTCATCTTAACCT-1</th>
      <td>data_5p</td>
      <td>3828</td>
      <td>1480</td>
      <td>1</td>
      <td>16</td>
      <td>CD8 T cell</td>
      <td>10X_5prime</td>
      <td>CD8 T cell</td>
    </tr>
    <tr>
      <th>data_5p-TTTGTCATCTTACCGC-1</th>
      <td>data_5p</td>
      <td>6444</td>
      <td>2388</td>
      <td>1</td>
      <td>28</td>
      <td>Plasmacytoid dendritic cell</td>
      <td>10X_5prime</td>
      <td>Plasmacytoid dendritic cell</td>
    </tr>
    <tr>
      <th>data_5p-TTTGTCATCTTGTTTG-1</th>
      <td>data_5p</td>
      <td>4457</td>
      <td>1662</td>
      <td>1</td>
      <td>11</td>
      <td>CD8 T cell</td>
      <td>10X_5prime</td>
      <td>CD8 T cell</td>
    </tr>
  </tbody>
</table>
<p>7378 rows × 8 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">celltype_batch2</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">adata2</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">item</span><span class="p">]</span> <span class="o">=</span> <span class="n">celltype_batch2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata2</span><span class="o">.</span><span class="n">obs_names</span><span class="p">][</span><span class="n">item</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 7378 × 33694
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="concatenate">
<h2>Concatenate<a class="headerlink" href="#concatenate" title="Link to this heading">#</a></h2>
<p>Now that we have loaded data from both the batches, we can combine them to create one anndata object.</p>
<p>For details on the parameters used for the function below, please see: <a class="reference external" href="https://anndata.readthedocs.io/en/latest/generated/anndata.concat.html">https://anndata.readthedocs.io/en/latest/generated/anndata.concat.html</a></p>
<p>For discussion/examples of concatenation, please see:
<a class="reference external" href="https://anndata.readthedocs.io/en/latest/concatenation.html">https://anndata.readthedocs.io/en/latest/concatenation.html</a></p>
<p>Special Note: Please pay attention to <code class="docutils literal notranslate"><span class="pre">join</span> <span class="pre">=</span> <span class="pre">'outer'</span></code> parameter. We are choosing to consider any gene that is expressed in either of the data. This means there may be some genes that are expressed in one data set but not in the other. What expression counts do these genes get in cells where they are not expressed? It depends on the input data matrix: if it is in sparse format they get 0, else NaN. So to ensure no NaN, we set <code class="docutils literal notranslate"><span class="pre">fill_value</span> <span class="pre">=</span> <span class="pre">0</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">anndata</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">],</span> 
                       <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>                        
                       <span class="n">join</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">,</span> 
                       <span class="n">label</span><span class="o">=</span><span class="s1">&#39;batch_id&#39;</span><span class="p">,</span> 
                       <span class="n">index_unique</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> 
                       <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


<span class="c1"># If you have more than 2 samples:</span>
<span class="c1">#adata = adata1_corr.concatenate(adata2_corr, adata3_corr, adata4_corr, </span>
<span class="c1">#                                axis = 0,</span>
<span class="c1">#                                join=&#39;outer&#39;, </span>
<span class="c1">#                                label=&#39;batch_id&#39;, </span>
<span class="c1">#                                index_unique=&#39;-&#39;, </span>
<span class="c1">#                                fill_value=0)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 33694
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;
</pre></div>
</div>
</div>
</div>
<p>Below we see the <code class="docutils literal notranslate"><span class="pre">.obs['batch_id']</span></code> of the adata and see that the first batch is called “0” and the second batch is called “1”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>data_3p-AAACCTGAGCATCATC-0-0    0
data_3p-AAACCTGAGCTAGTGG-0-0    0
data_3p-AAACCTGCACATTAGC-0-0    0
data_3p-AAACCTGCACTGTTAG-0-0    0
data_3p-AAACCTGCATAGTAAG-0-0    0
                               ..
data_5p-TTTGTCATCCACGTTC-1-1    1
data_5p-TTTGTCATCGCGTAGC-1-1    1
data_5p-TTTGTCATCTTAACCT-1-1    1
data_5p-TTTGTCATCTTACCGC-1-1    1
data_5p-TTTGTCATCTTGTTTG-1-1    1
Name: batch_id, Length: 15476, dtype: category
Categories (2, object): [&#39;0&#39;, &#39;1&#39;]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">X</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># check for NaN:</span>
<span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0
</pre></div>
</div>
</div>
</div>
<p>As seen above, there are no NaN in the combined data, so we can proceed normally.</p>
</section>
<section id="basic-qc-and-normalization">
<h2>Basic QC and normalization<a class="headerlink" href="#basic-qc-and-normalization" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Distribution of library size (total counts); number of genes per cell</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_total_counts&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_total_counts&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-1&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log-library size&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_n_genes_by_counts&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_n_genes_by_counts&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-1&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log-N genes per cell&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;matplotlib.legend.Legend at 0x7f0090d0cb20&gt;
</pre></div>
</div>
<img alt="_images/c115292299045c26ad9e72660291af9b10f4b8bd9e1d6f9d3a023b5f55656e0d.png" src="_images/c115292299045c26ad9e72660291af9b10f4b8bd9e1d6f9d3a023b5f55656e0d.png" />
</div>
</div>
<p>If you have more than 2 batch, you can run a for loop to visualize the distribution of each batch compared to the whole data. For this, you can adapt the following example code:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Distribution of library size (total counts)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">1</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">])):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_total_counts&#39;</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;All cells&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;log1p_total_counts&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">item</span><span class="p">],</span> <span class="mi">100</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-&#39;</span> <span class="o">+</span> <span class="n">item</span><span class="p">,</span> 
            <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log-library size&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/407084b77ac2d6fc6a322650334b66d7160ab71f6b7ff6a415d73ee949ba8fff.png" src="_images/407084b77ac2d6fc6a322650334b66d7160ab71f6b7ff6a415d73ee949ba8fff.png" />
</div>
</div>
<p>We see that on average batch-1 cells have higher library size and higher number of genes expressed per cell. We will keep this mind as we analyze the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Distribution of number of genes a cell is expressed in</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;n_cells_by_counts&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log-N cells per gene&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;n_cells_by_counts&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Log-N cells per gene&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency, trimmed axis&#39;</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(0.0, 500.0)
</pre></div>
</div>
<img alt="_images/dad58c321fc68638fee44b59e7639cb975687840f75c2049b30c7c36deb12988.png" src="_images/dad58c321fc68638fee44b59e7639cb975687840f75c2049b30c7c36deb12988.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove genes expressed in less than 64 cells:</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">6</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Normalizing + log transformation</span>

<span class="c1"># Store raw counts for future</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;raw_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Normalize with median library size</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="c1"># Store the normalized counts for future</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;norm_counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># Take log2 of the normalized counts</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="highly-variable-genes-pca-nearest-neighbor-graph-umap-fdl-phenograph">
<h2>Highly variable genes, PCA, Nearest Neighbor Graph, UMAP, FDL, Phenograph<a class="headerlink" href="#highly-variable-genes-pca-nearest-neighbor-graph-umap-fdl-phenograph" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># HVG</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layer</span> <span class="o">=</span> <span class="s1">&#39;raw_counts&#39;</span><span class="p">,</span> <span class="n">n_top_genes</span> <span class="o">=</span> <span class="mi">4000</span><span class="p">,</span> <span class="n">flavor</span> <span class="o">=</span> <span class="s1">&#39;seurat_v3&#39;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;id_hvg&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;pca&#39;</span><span class="p">][</span><span class="s1">&#39;variance_ratio&#39;</span><span class="p">])),</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;pca&#39;</span><span class="p">][</span><span class="s1">&#39;variance_ratio&#39;</span><span class="p">])</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="s1">&#39;.-&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Principal Component&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;% Variance Explained&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0, 0.5, &#39;% Variance Explained&#39;)
</pre></div>
</div>
<img alt="_images/15f49e8cb4d0a8d1d836757cf71fafb9d3edf32a43d17d2650df357f18bfd237.png" src="_images/15f49e8cb4d0a8d1d836757cf71fafb9d3edf32a43d17d2650df357f18bfd237.png" />
</div>
</div>
<p>Based on the plot above, we will select 30 principal components.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">30</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;neighbors_30&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clustering using PhenoGraph</span>
<span class="n">sc</span><span class="o">.</span><span class="n">external</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">phenograph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">clustering_algo</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">jaccard</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">primary_metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                          <span class="n">resolution_parameter</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">scipy</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;pheno_jaccard_ig&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s1">&#39;pheno_jaccard_ig&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Finding 30 nearest neighbors using minkowski metric and &#39;auto&#39; algorithm
Neighbors computed in 7.513487100601196 seconds
Jaccard graph constructed in 14.946049213409424 seconds
Running Leiden optimization
Leiden completed in 0.4645841121673584 seconds
Sorting communities by size, please wait ...
PhenoGraph completed in 33.102680683135986 seconds
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualize-the-results">
<h2>Visualize the results<a class="headerlink" href="#visualize-the-results" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">,</span> <span class="s1">&#39;pheno_leiden&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="_images/6864074d449f13e520b3f5ad4e872f35f4dbe164faa7dbdac5a802646c17c501.png" src="_images/6864074d449f13e520b3f5ad4e872f35f4dbe164faa7dbdac5a802646c17c501.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">,</span> <span class="s1">&#39;pheno_leiden&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<img alt="_images/1ff78a89346463883357d3cc2a088cc3d3904a347bc11bfe249686e6475ef463.png" src="_images/1ff78a89346463883357d3cc2a088cc3d3904a347bc11bfe249686e6475ef463.png" />
</div>
</div>
</section>
<section id="visualize-some-genes">
<h2>Visualize some genes<a class="headerlink" href="#visualize-some-genes" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CD3E&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD4&#39;</span><span class="p">,</span> <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span> <span class="s1">&#39;MS4A1&#39;</span><span class="p">,</span> <span class="s1">&#39;CD19&#39;</span><span class="p">,</span> <span class="s1">&#39;LYZ&#39;</span><span class="p">,</span> <span class="s1">&#39;FCGR3A&#39;</span><span class="p">,</span> <span class="s1">&#39;NKG7&#39;</span><span class="p">,</span> <span class="s1">&#39;CST3&#39;</span><span class="p">,</span> <span class="s1">&#39;PPBP&#39;</span><span class="p">,</span> <span class="s1">&#39;PTPRC&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mf">1.5</span><span class="p">))</span>
<span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ac</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">groupby</span> <span class="o">=</span> <span class="s1">&#39;pheno_leiden&#39;</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                   <span class="n">layer</span> <span class="o">=</span> <span class="s1">&#39;norm_counts&#39;</span><span class="p">,</span> 
                   <span class="n">show</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_dotplot.py:747: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39;, &#39;norm&#39; will be ignored
  dot_ax.scatter(x, y, **kwds)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/95c633fd0c10b95cd534863b3fb3e0155a9e9b9671c792c2573cf9ae5d8686da.png"><img alt="_images/95c633fd0c10b95cd534863b3fb3e0155a9e9b9671c792c2573cf9ae5d8686da.png" src="_images/95c633fd0c10b95cd534863b3fb3e0155a9e9b9671c792c2573cf9ae5d8686da.png" style="width: 1038px; height: 623px;" /></a>
</div>
</div>
<section id="highlight-b-cells">
<h3>Highlight B cells<a class="headerlink" href="#highlight-b-cells" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bcells</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pheno_leiden&#39;</span><span class="p">],</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">22</span><span class="p">]))[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">1</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
           <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> 
           <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">markerscale</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Colored by batch-id&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> 
           <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;All cells&#39;</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">bcells</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">bcells</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> 
            <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;B cells&#39;</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">markerscale</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Highlighting B-cells&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;Highlighting B-cells&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/fe8b8db47f2d1dd0d773321d4f3e5908d76bb80977cea25531965c824eee1bf4.png"><img alt="_images/fe8b8db47f2d1dd0d773321d4f3e5908d76bb80977cea25531965c824eee1bf4.png" src="_images/fe8b8db47f2d1dd0d773321d4f3e5908d76bb80977cea25531965c824eee1bf4.png" style="width: 1014px; height: 398px;" /></a>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="batch-effect-correction">
<h1>Batch Effect Correction<a class="headerlink" href="#batch-effect-correction" title="Link to this heading">#</a></h1>
<p>Batch effect is a technical effect that can confound biological interpretations. We mostly design experiments in such a way that there is as little batch effect as possible. But often times, we wish to compare our data at hand against some other published data. It is natural to expect some batch effect in such cases. Therefore, we need to find computational solutions to correct such technical confounder. As such, this is one of the busiest areas for computational methods development.</p>
<p>While there exists many methods to perform such correction, there is a common underlying assumption in all the methods: there must be some sub-populations with the same phenotype (cell type or state) across batches. Under this assumption, a majority of the methods begin by computing nearest neighbor cell or mutually nearest neighbor (MNN) between batches to identify the similar cells. While we have covered nearest neighbor cell in the previous sessions, we have not discussed mutually nearest neighbor cells. We define mutually nearest neighbor cells as follows: If a pair of cells from each batch is contained in each other’s set of nearest neighbors, those cells are considered to be mutual nearest neighbors. The methods interpret these pairs as consisting cells that belong to the same cell type or state despite being generated in different batches.</p>
<p>Here, we will discuss some of the most popular methods for batch effect correction.</p>
<p>Scanorama: <a class="github reference external" href="https://github.com/brianhie/scanorama">brianhie/scanorama</a></p>
<ul class="simple">
<li><p>Scanorama uses randomized singular value decomposition (SVD) - similar to PCA - to compress the gene expression profiles into a low-dimensional embedding.</p></li>
<li><p>It then searches for MNNs in the dimensionality reduced spaces and uses them in a careful way (weighted by similarity) to guide batch integration.</p></li>
<li><p>Note: Scanorama also corrects the expression of the genes in all the batches (i.e. outputs a batch corrected gene expression matrix) that can be utilized downstream for further analysis such as visualization, differential expression etc.</p></li>
</ul>
<p>FastMNN:</p>
<ul class="simple">
<li><p>FastMNN uses principal component analysis (PCA) to compress the gene expression profiles into a low-dimensional embedding.</p></li>
<li><p>It then searches for MNNs in the dimensionality reduced spaces and uses them to compute a translation vector to align the datasets into a shared space.</p></li>
<li><p>More details can be found here: <a class="reference external" href="https://marionilab.github.io/FurtherMNN2018/theory/description.html">https://marionilab.github.io/FurtherMNN2018/theory/description.html</a></p></li>
<li><p>Note: The original version is <code class="docutils literal notranslate"><span class="pre">mnnCorrect</span></code>, but it is more computationally demanding in terms of CPU and memory as it uses the full high dimensional gene expression space. (<a class="reference external" href="https://rdrr.io/bioc/batchelor/man/mnnCorrect.html">https://rdrr.io/bioc/batchelor/man/mnnCorrect.html</a>)</p></li>
<li><p>Note: FastMNN also corrects the expression of the genes in all the batches (i.e. outputs a batch corrected gene expression matrix) that can be utilized downstream for further analysis such as visualization, differential expression etc.</p></li>
</ul>
<p>Harmony: <a class="reference external" href="https://portals.broadinstitute.org/harmony/articles/quickstart.html">https://portals.broadinstitute.org/harmony/articles/quickstart.html</a></p>
<ul class="simple">
<li><p>Harmony uses Principal Component Analysis (PCA) to compress the gene expression profiles into a low-dimensional embedding.</p></li>
<li><p>In the PCA space, Harmony iteratively removes batch effects present. At each iteration, it clusters similar cells from different batches and maximizes the diversity of batches within each cluster. It then calculates a correction factor for each cell to be applied. It is good for visualization.</p></li>
<li><p>Note: Harmony <em>does not</em> correct the expression of the genes in all the batches.</p></li>
<li><p>Note: Harmony has also been noted to mess up the structure of the data post-correction, so we recommend to use with caution and ensuring that biological interpretation remains consistent.</p></li>
</ul>
<p>Seurat Integrate: <a class="reference external" href="https://satijalab.org/seurat/articles/integration_introduction.html">https://satijalab.org/seurat/articles/integration_introduction.html</a></p>
<ul class="simple">
<li><p>Seurate Integrate uses canonical correlation analysis (CCA) to compress the gene expression profiles into a low-dimensional embedding.</p></li>
<li><p>It then searches for MNNs in the dimensionally reduced spaces and uses them as “anchors” to guide batch correction.</p></li>
<li><p>Note: Seurat Integrate <em>does not</em> correct the expression of the genes in all the batches.</p></li>
</ul>
<p>We provide example code to run four different batch correction methods.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="batch-effect-correction-scanorama">
<h1>Batch effect correction - Scanorama<a class="headerlink" href="#batch-effect-correction-scanorama" title="Link to this heading">#</a></h1>
<p>While Scanorama is integrated into Scanpy, the current implementation on Scanpy does not correct the gene expression matrix. Therefore we will use the original stand-alone implementation of Scanorama (i.e. we will not call the function via Scanpy).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanorama</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;
</pre></div>
</div>
</div>
</div>
<p>First, we need to divide the data into respective batches.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Our batch variable is stored under &#39;batch&#39;</span>
<span class="n">batch_id</span> <span class="o">=</span> <span class="s1">&#39;batch&#39;</span>

<span class="n">adatas</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_id</span><span class="p">]):</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_id</span><span class="p">]</span> <span class="o">==</span> <span class="n">c</span>
    <span class="n">adatas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="p">:])</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can run Scanorama to obtain corrected expression counts.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">corrected</span> <span class="o">=</span> <span class="n">scanorama</span><span class="o">.</span><span class="n">correct_scanpy</span><span class="p">(</span><span class="n">adatas</span><span class="p">,</span> 
                                     <span class="n">return_dimred</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                     <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                                     <span class="n">dimred</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> 
                                     <span class="n">knn</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Found 13386 genes among all datasets
Processing datasets (0, 1)
</pre></div>
</div>
</div>
</div>
<p>A note on the code above (you can also view the help page for scanorama):</p>
<ol class="arabic simple">
<li><p>The first input is a list of the adata, or batches we want to correct. Note: Scanorama directly acts on <code class="docutils literal notranslate"><span class="pre">.X</span></code>.</p></li>
<li><p>return_dimred = True implies that the method will return the reduced dimensions (like PCA) where the two batches have been co-embedded or co-aligned.</p></li>
<li><p>verbose = True indicates whether to output the intermediate comments.</p></li>
<li><p>dimred = 30 is the number of dimensions on which Scanorama will perform the cell nearest neighbor mathching. I set it to 30 because based on our analysis above, we find that 30 principal components capture enough variance.</p></li>
<li><p>knn = 30 is the number of nearest neighbors to estimate. This number should be smaller if you know the size of your sub-populations that are shared across batches is smaller.</p></li>
</ol>
<p>Note: There is another parameter called <code class="docutils literal notranslate"><span class="pre">hvg</span></code> to specify the number of highly variable genes (automatically estimated by scanorama), which we welcome you to play with. The only catch is that if you specify this number, the output matrix will only be corrected with highly varying genes (i.e. not all the genes).</p>
<p>The output is stored in <code class="docutils literal notranslate"><span class="pre">corrected</span></code> which stores output <code class="docutils literal notranslate"><span class="pre">anndata</span></code> as lists. For example, <code class="docutils literal notranslate"><span class="pre">corrected[0]</span></code> contains the corrected data for batch-0 and <code class="docutils literal notranslate"><span class="pre">corrected[1]</span></code> contains the corrected data for batch-1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">corrected</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 8098 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;X_scanorama&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">corrected</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 7378 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;X_scanorama&#39;
</pre></div>
</div>
</div>
</div>
<p>We can combine the results and create a new anndata called adata_corrected.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_corrected</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">corrected</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_corrected</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;X_scanorama&#39;
</pre></div>
</div>
</div>
</div>
<p>Note: In the above combined adata_corrected data, we have X_pca, X_umap and X_draw_graph_fa which were computed previously and should not used for the corrected combined data. So we need to recompute them. But first let’s save them as something else so they don’t overwritten.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span>
<span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save for later:</span>
<span class="c1"># adata_corrected.write_h5ad(&#39;/path/to/store/batch_data_scanorama_corrected.h5ad&#39;)</span>
</pre></div>
</div>
</div>
</div>
<section id="nearest-neighbors-umap-fdl">
<h2>Nearest neighbors, UMAP, FDL<a class="headerlink" href="#nearest-neighbors-umap-fdl" title="Link to this heading">#</a></h2>
<p>Note: Corrected reduced dimensions - PCA analogous - have already been computed (they are stored in <code class="docutils literal notranslate"><span class="pre">.obsm['X_scanorama']</span></code>) so all we need to do is run nearest neighbors algorithm followed by UMAP/FDL for visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on corrected PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_corrected</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_scanorama&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;neighbors_30_corrected&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_corrected</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30_corrected&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata_corrected</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;neighbors_30_corrected&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualize">
<h2>Visualize<a class="headerlink" href="#visualize" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of un-corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_corrected</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of un-corrected data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;FDL of un-corrected data&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/4c276648a6a4d7d219a198845687edf1357c5003246f10bd7c914e41f0692407.png"><img alt="_images/4c276648a6a4d7d219a198845687edf1357c5003246f10bd7c914e41f0692407.png" src="_images/4c276648a6a4d7d219a198845687edf1357c5003246f10bd7c914e41f0692407.png" style="width: 1014px; height: 763px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">help</span><span class="p">(</span><span class="n">scanorama</span><span class="o">.</span><span class="n">correct</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Help on function correct in module scanorama.scanorama:

correct(datasets_full, genes_list, return_dimred=False, batch_size=5000, verbose=2, ds_names=None, dimred=100, approx=True, sigma=15, alpha=0.1, knn=20, return_dense=False, hvg=None, union=False, seed=0)
    Integrate and batch correct a list of data sets.
    
    Parameters
    ----------
    datasets_full : `list` of `scipy.sparse.csr_matrix` or of `numpy.ndarray`
        Data sets to integrate and correct.
    genes_list: `list` of `list` of `string`
        List of genes for each data set.
    return_dimred: `bool`, optional (default: `False`)
        In addition to returning batch corrected matrices, also returns
        integrated low-dimesional embeddings.
    batch_size: `int`, optional (default: `5000`)
        The batch size used in the alignment vector computation. Useful when
        correcting very large (&gt;100k samples) data sets. Set to large value
        that runs within available memory.
    verbose: `bool` or `int`, optional (default: 2)
        When `True` or not equal to 0, prints logging output.
    ds_names: `list` of `string`, optional
        When `verbose=True`, reports data set names in logging output.
    dimred: `int`, optional (default: 100)
        Dimensionality of integrated embedding.
    approx: `bool`, optional (default: `True`)
        Use approximate nearest neighbors, greatly speeds up matching runtime.
    sigma: `float`, optional (default: 15)
        Correction smoothing parameter on Gaussian kernel.
    alpha: `float`, optional (default: 0.10)
        Alignment score minimum cutoff.
    knn: `int`, optional (default: 20)
        Number of nearest neighbors to use for matching.
    return_dense: `bool`, optional (default: `False`)
        Return `numpy.ndarray` matrices instead of `scipy.sparse.csr_matrix`.
    hvg: `int`, optional (default: None)
        Use this number of top highly variable genes based on dispersion.
    seed: `int`, optional (default: 0)
        Random seed to use.
    
    Returns
    -------
    corrected, genes
        By default (`return_dimred=False`), returns a two-tuple containing a
        list of `scipy.sparse.csr_matrix` each with batch corrected values,
        and a single list of genes containing the intersection of inputted
        genes.
    
    integrated, corrected, genes
        When `return_dimred=True`, returns a three-tuple containing a list
        of `numpy.ndarray` with integrated low dimensional embeddings, a list
        of `scipy.sparse.csr_matrix` each with batch corrected values, and a
        a single list of genes containing the intersection of inputted genes.
</pre></div>
</div>
</div>
</div>
</section>
<section id="true-celltypes">
<h2>True celltypes<a class="headerlink" href="#true-celltypes" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_corrected</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">,</span> <span class="s1">&#39;CellType&#39;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/e5d9e834f6bf82459dc6cc0d4d6255a6ba0aff6e79f21deccde2c0adfe90768b.png"><img alt="_images/e5d9e834f6bf82459dc6cc0d4d6255a6ba0aff6e79f21deccde2c0adfe90768b.png" src="_images/e5d9e834f6bf82459dc6cc0d4d6255a6ba0aff6e79f21deccde2c0adfe90768b.png" style="width: 994px; height: 299px;" /></a>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="batch-effect-correction-harmony">
<h1>Batch effect correction - Harmony<a class="headerlink" href="#batch-effect-correction-harmony" title="Link to this heading">#</a></h1>
<p>Harmony is another popular method for correcting batch effect in single-cell RNA-sequencing data. As discussed in the beginning, Harmony iteratively removes batch effects in the PCA space. In each iteration, it clusters similar cells from different batches and maximizing the diversity of batches within each cluster; it then calculates a correction factor for each cell to be applied. While it is a fast method, there are a couple caveats: 1) Since it operates only on the PCA space, any non-linear association between features could be missed or altered; and 2) Harmony does not provide a corrected gene expression matrix. Therefore, the output of Harmony can only be used for visualization and clustering but not for differential expression analysis or any gene centric analysis, since one does not get a corrected expression matrix back. One would need to be careful to perform downstream analysis with Harmony.</p>
<p>Harmony is implemented in Scanpy: <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.external.pp.harmony_integrate.html">https://scanpy.readthedocs.io/en/stable/generated/scanpy.external.pp.harmony_integrate.html</a>.</p>
<p>Let’s first make a copy of <code class="docutils literal notranslate"><span class="pre">adata</span></code> for application of Harmony. Note we do not simply do <code class="docutils literal notranslate"><span class="pre">adata_harmony</span> <span class="pre">=</span> <span class="pre">adata</span></code> but specify <code class="docutils literal notranslate"><span class="pre">adata_harmony</span> <span class="pre">=</span> <span class="pre">adata.copy()</span></code> because of the issue of mutable variables in Python. See more here: <a class="reference external" href="https://www.geeksforgeeks.org/mutable-vs-immutable-objects-in-python/">https://www.geeksforgeeks.org/mutable-vs-immutable-objects-in-python/</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_harmony</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Please run this step if we did not install harmonypy</span>
<span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>harmonypy
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Collecting harmonypy
  Downloading harmonypy-0.0.9-py3-none-any.whl.metadata (3.7 kB)
Requirement already satisfied: pandas in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from harmonypy) (2.0.3)
Requirement already satisfied: numpy in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from harmonypy) (1.24.4)
Requirement already satisfied: scipy in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from harmonypy) (1.10.1)
Requirement already satisfied: scikit-learn in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from harmonypy) (1.1.3)
Requirement already satisfied: python-dateutil&gt;=2.8.2 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from pandas-&gt;harmonypy) (2.8.2)
Requirement already satisfied: pytz&gt;=2020.1 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from pandas-&gt;harmonypy) (2023.3.post1)
Requirement already satisfied: tzdata&gt;=2022.1 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from pandas-&gt;harmonypy) (2023.4)
Requirement already satisfied: joblib&gt;=1.0.0 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from scikit-learn-&gt;harmonypy) (1.3.2)
Requirement already satisfied: threadpoolctl&gt;=2.0.0 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from scikit-learn-&gt;harmonypy) (3.2.0)
Requirement already satisfied: six&gt;=1.5 in /lilac/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages (from python-dateutil&gt;=2.8.2-&gt;pandas-&gt;harmonypy) (1.16.0)
Downloading harmonypy-0.0.9-py3-none-any.whl (20 kB)
Installing collected packages: harmonypy
Successfully installed harmonypy-0.0.9
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">external</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">harmony_integrate</span><span class="p">(</span><span class="n">adata_harmony</span><span class="p">,</span> 
                                 <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;batch_id&#39;</span><span class="p">,</span> 
                                 <span class="n">basis</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span> 
                                 <span class="n">adjusted_basis</span><span class="o">=</span><span class="s1">&#39;X_pca_harmony&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2024-03-13 14:53:55,934 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...
2024-03-13 14:53:59,890 - harmonypy - INFO - sklearn.KMeans initialization complete.
2024-03-13 14:53:59,948 - harmonypy - INFO - Iteration 1 of 10
2024-03-13 14:54:03,077 - harmonypy - INFO - Iteration 2 of 10
2024-03-13 14:54:06,157 - harmonypy - INFO - Iteration 3 of 10
2024-03-13 14:54:09,210 - harmonypy - INFO - Iteration 4 of 10
2024-03-13 14:54:12,279 - harmonypy - INFO - Iteration 5 of 10
2024-03-13 14:54:13,657 - harmonypy - INFO - Iteration 6 of 10
2024-03-13 14:54:16,795 - harmonypy - INFO - Iteration 7 of 10
2024-03-13 14:54:19,865 - harmonypy - INFO - Converged after 7 iterations
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_harmony</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;X_pca_harmony&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;
</pre></div>
</div>
</div>
</div>
<p>Now we just need to recompute the neighbors, UMAP and FDL on the above computed X_pca_harmony:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Save to prevent overwriting</span>
<span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span>
<span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on PCA_harmony</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_harmony</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca_harmony&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;neighbors_30_harmony&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_harmony</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30_harmony&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata_harmony</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;neighbors_30_harmony&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of Harmony corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_UMAP&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of un-corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of Harmony corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata_harmony</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;old_FDL&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of un-corrected data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;FDL of un-corrected data&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/9bfe69fb85e2582fb840afe08c81ab3bd59198bd2dc11f0020f3e356fc1bcabf.png"><img alt="_images/9bfe69fb85e2582fb840afe08c81ab3bd59198bd2dc11f0020f3e356fc1bcabf.png" src="_images/9bfe69fb85e2582fb840afe08c81ab3bd59198bd2dc11f0020f3e356fc1bcabf.png" style="width: 1014px; height: 763px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_harmony</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CellType&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/f980babf606176694a34eaf8b8d2cf2e3004dc4f398c41348748fbe9563955c7.png"><img alt="_images/f980babf606176694a34eaf8b8d2cf2e3004dc4f398c41348748fbe9563955c7.png" src="_images/f980babf606176694a34eaf8b8d2cf2e3004dc4f398c41348748fbe9563955c7.png" style="width: 508px; height: 296px;" /></a>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="batch-effect-correction-fastmnn">
<h1>Batch effect correction - FastMNN<a class="headerlink" href="#batch-effect-correction-fastmnn" title="Link to this heading">#</a></h1>
<p>Fast Mutual Nearest Neighbors based batch effect correction was the first batch effect correction method proposed for scRNA-seq data (<a class="reference external" href="https://www.nature.com/articles/nbt.4091">https://www.nature.com/articles/nbt.4091</a>). The idea of FastMNN is to first find mutually nearest neighbors across batched, assume that mutually nearest neighbors mean they have the same cell phenotype (even though they come from different batches), compute the difference in expression between the mutually nearest neighbors and aggregate those differences across all similar phenotypes as a correction vector for the data to correct for the batch effect.</p>
<p>Since FastMNN was written in R, we will need to do some brief preparations. Below, we provide code for running FastMNN in two ways: 1) On highly varying genes (HVG) only, which uses HVG to construct MNN and takes less time/memory to run 2) All genes, which considers all the genes but takes more time/memory to run. We recommend using the HVG approach not only for speed but using all genes may introduce additional noise on top of biological signal (which we assume is captured by HVG).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;
</pre></div>
</div>
</div>
</div>
<section id="implementation-with-hvgs">
<h2>Implementation with HVGs<a class="headerlink" href="#implementation-with-hvgs" title="Link to this heading">#</a></h2>
<p>If you have previously calculated HVGs using <code class="docutils literal notranslate"><span class="pre">sc.pp.highly_variable_genes()</span></code>, you can use the HVGs as the features for batch correction. That’s what we do here. However you can imagine that there might be other ways of computing HVGs, for example you can compute HVG for each batch separately and take the union (or intersection) as the final HVG.</p>
<p>First we prepare the input to FastMNN: It simply is an anndata object that has the log-normalized data in <code class="docutils literal notranslate"><span class="pre">.X</span></code>,  contains information for which batch each cell belongs to in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> and gene names in <code class="docutils literal notranslate"><span class="pre">.var</span></code>. Since we are interested in running the correction using HVG, we will also save the HVG as a csv file to later load into R.</p>
<p>As an alternate to our previous attempts to run R from Python, we will first write a .R script that contains the R code we want to run and then run the .R script file. This is necessary for fastMNN and Seurat Integration as the output produced by these methods is not yet compatible with “automatic” loading into Python. So, we will first save the data as a h5ad file, load it in R, run the R code, save the output as h5ad file and load it back into Python.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Prepare data</span>
<span class="n">fastmnn_input</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> 
                           <span class="n">obs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;batch&#39;</span><span class="p">]],</span>
                           <span class="n">var</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>

<span class="n">hvg</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hvg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>HES4</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ISG15</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AGRN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TNFRSF18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TNFRSF4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>3995</th>
      <td>MT-ND4L</td>
    </tr>
    <tr>
      <th>3996</th>
      <td>MT-ND5</td>
    </tr>
    <tr>
      <th>3997</th>
      <td>MT-ND6</td>
    </tr>
    <tr>
      <th>3998</th>
      <td>MT-CYB</td>
    </tr>
    <tr>
      <th>3999</th>
      <td>AC004556.1</td>
    </tr>
  </tbody>
</table>
<p>4000 rows × 1 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i fastmnn_input -i hvg
fastmnn_input
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 13386 15476 
metadata(0):
assays(1): X
rownames(13386): FO538757.2 AP006222.2 ... AC004556.1 AC240274.1
rowData names(0):
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(1): batch
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
head(hvg)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>         0
0     HES4
1    ISG15
2     AGRN
3 TNFRSF18
4  TNFRSF4
5    MXRA8
</pre></div>
</div>
</div>
</div>
<p>We need to adjust the <code class="docutils literal notranslate"><span class="pre">X</span></code> assay in fastmnn_input as the <code class="docutils literal notranslate"><span class="pre">logcounts</span></code> assay:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# store the data as logcounts for fastMNN to identify it
logcounts(fastmnn_input) &lt;- assay(fastmnn_input, &quot;X&quot;)
print(&#39;logcounts done&#39;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;logcounts done&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
unlist(hvg)[1:5]
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>        01         02         03         04         05 
    &quot;HES4&quot;    &quot;ISG15&quot;     &quot;AGRN&quot; &quot;TNFRSF18&quot;  &quot;TNFRSF4&quot; 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# test to see if this way of extract HVG works
fastmnn_input[unlist(hvg), 1:5]
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 4000 5 
metadata(0):
assays(2): X logcounts
rownames(4000): HES4 ISG15 ... MT-CYB AC004556.1
rowData names(0):
colnames(5): data_3p-AAACCTGAGCATCATC-0-0 data_3p-AAACCTGAGCTAGTGG-0-0
  data_3p-AAACCTGCACATTAGC-0-0 data_3p-AAACCTGCACTGTTAG-0-0
  data_3p-AAACCTGCATAGTAAG-0-0
colData names(1): batch
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p>Run fastMNN correct:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
library(batchelor, quietly=TRUE)
out &lt;- fastMNN(fastmnn_input, 
               batch = fastmnn_input$batch, 
               subset.row = unlist(hvg), 
               correct.all = TRUE, 
               d = 30, 
               k = 30)
</pre></div>
</div>
</div>
</div>
<p>Some comments on the code above:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastmnn_input</span></code> is the object (<code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code>) which stores the data on which we want to do batch correction</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">batch</span> <span class="pre">=</span> <span class="pre">fastmnn_input$batch</span></code> specifies the batches we want to correct</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subset.row</span> <span class="pre">=</span> <span class="pre">hvg</span></code> indicates that we want to use hvg only to do PCA, mnn etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">correct.all</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> indicates that we want to correct expression of all genes (even though we only used hvg for pca etc.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">d</span> <span class="pre">=</span> <span class="pre">30</span></code> indicates the number of lower dimensional spaces where the two batches will be embedded</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">30</span></code> indicates the number of mutually nearest neighbors to compute</p></li>
</ul>
<section id="process-the-corrected-output">
<h3>Process the corrected output<a class="headerlink" href="#process-the-corrected-output" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
out
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 13386 15476 
metadata(2): merge.info pca.info
assays(1): reconstructed
rownames(13386): FO538757.2 AP006222.2 ... AC004556.1 AC240274.1
rowData names(1): rotation
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(1): batch
reducedDimNames(1): corrected
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p>It is difficult to store the format of data fastMNN produces so we manipulate it to be of different class and call it back to Python.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o out2
mat &lt;- assay(out, &quot;reconstructed&quot;)
mat &lt;- as(mat, &quot;dgCMatrix&quot;)
out2 &lt;- SingleCellExperiment(list(counts = mat))
rownames(out2) &lt;- rownames(out)
colnames(out2) &lt;- colnames(out)
reducedDim(out2) &lt;- reducedDim(out)
reducedDimNames(out2) &lt;- &#39;corrected_pca&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">out2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obsm: &#39;corrected_pca&#39;
</pre></div>
</div>
</div>
</div>
<p>Note: As an output we get <code class="docutils literal notranslate"><span class="pre">out2</span></code>, which has the corrected expression in <code class="docutils literal notranslate"><span class="pre">.X</span></code> and the corrected pca in <code class="docutils literal notranslate"><span class="pre">.obsm</span></code>. We can append these to our <code class="docutils literal notranslate"><span class="pre">adata</span></code> and work from there. However, please note that fastMNN produces a lot of other outputs than just these. It also gives you a rotation matrix in <code class="docutils literal notranslate"><span class="pre">rowData</span></code> which stores the matrix that was used to rotate the PCA. If you are interested in these outputs, please extract them manually, for example: <code class="docutils literal notranslate"><span class="pre">mat2</span> <span class="pre">&lt;-</span> <span class="pre">as(rowData(out),</span> <span class="pre">'data.frame')</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure that obs_names are in the same order in out2 and adata:</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="n">out2</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not in right order&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure that var_names are in the same order in out2 and adata:</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="n">out2</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not in right order&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;fastMNN_hvg_corrected&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out2</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;fastMNN_hvg_pca&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;corrected_pca&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;fastMNN_hvg_pca&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;, &#39;fastMNN_hvg_corrected&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Avoid overwriting uncorrected UMAP and FDL</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id1">
<h3>Nearest neighbors, UMAP, FDL<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;fastMNN_hvg_pca&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;fastMNN_hvg&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;fastMNN_hvg&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;fastMNN_hvg&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="visualize-the-data">
<h3>Visualize the data<a class="headerlink" href="#visualize-the-data" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of FastMNN corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of un-corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of FastMNN corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of un-corrected data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;FDL of un-corrected data&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/43bfba27e0e96979415483e4ecb1f9156d8368c3d39f31c491059d72a3e28b22.png"><img alt="_images/43bfba27e0e96979415483e4ecb1f9156d8368c3d39f31c491059d72a3e28b22.png" src="_images/43bfba27e0e96979415483e4ecb1f9156d8368c3d39f31c491059d72a3e28b22.png" style="width: 1014px; height: 763px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CellType&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/16ecc8fecbdfbc05fbaa8cbee453f2fd762b3fe74d46b7fab247b838ff0bc45f.png"><img alt="_images/16ecc8fecbdfbc05fbaa8cbee453f2fd762b3fe74d46b7fab247b838ff0bc45f.png" src="_images/16ecc8fecbdfbc05fbaa8cbee453f2fd762b3fe74d46b7fab247b838ff0bc45f.png" style="width: 508px; height: 296px;" /></a>
</div>
</div>
</section>
</section>
<section id="implementation-with-all-genes">
<h2>Implementation with all genes<a class="headerlink" href="#implementation-with-all-genes" title="Link to this heading">#</a></h2>
<p>We now provide the code to do batch correction using all the genes. This may take much longer time than above (and your laptop may run out of memory, if you are running it on your laptop), and may only be beneficial when the selected HVG are not sufficient to capture cell type similarity across data sets.</p>
<p>First we prepare the input to FastMNN: It simply is an anndata object that has the log-normalized data in .X and contains information for which batch each cell belongs to in .obs (and ofcourse has gene names in .var).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fastmnn_input_all_genes</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> 
                 <span class="n">obs</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s1">&#39;batch&#39;</span><span class="p">]],</span>
                 <span class="n">var</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fastmnn_input_all_genes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;batch&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i fastmnn_input_all_genes 
fastmnn_input_all_genes
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 13386 15476 
metadata(0):
assays(1): X
rownames(13386): FO538757.2 AP006222.2 ... AC004556.1 AC240274.1
rowData names(0):
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(1): batch
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p>We need to adjust the <code class="docutils literal notranslate"><span class="pre">X</span></code> assay in fastmnn_input as the <code class="docutils literal notranslate"><span class="pre">logcounts</span></code> assay:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# store the data as logcounts for fastMNN to identify it
logcounts(fastmnn_input_all_genes) &lt;- assay(fastmnn_input_all_genes, &quot;X&quot;)
print(&#39;logcounts done&#39;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;logcounts done&quot;
</pre></div>
</div>
</div>
</div>
<p>Run fastMNN correct:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
library(batchelor, quietly=TRUE)
out_all_genes &lt;- fastMNN(fastmnn_input_all_genes, 
                         batch = fastmnn_input_all_genes$batch, 
                         correct.all = TRUE, 
                         d = 30, 
                         k = 30)
</pre></div>
</div>
</div>
</div>
<p>Some comments on the code above:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastmnn_input_all_genes</span></code> is the object (<code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code>) which stores the data on which we want to do batch correction</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">batch</span> <span class="pre">=</span> <span class="pre">fastmnn_input_all_genes$batch</span></code> specifies the batches we want to correct</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">correct.all</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> indicates that we want to correct expression of all genes (even though we only used hvg for pca etc.)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">d</span> <span class="pre">=</span> <span class="pre">30</span></code> indicates the number of lower dimensional spaces where the two batches will be embedded</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">k</span> <span class="pre">=</span> <span class="pre">30</span></code> indicates the number of mutually nearest neighbors to compute</p></li>
</ul>
<p>Same as in the case with HVG we will extract only the needed output and for the rest, please extract them manually (for example using sample code provided above in the case of HVG).</p>
<section id="id2">
<h3>Process the corrected output<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
out_all_genes
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 13386 15476 
metadata(2): merge.info pca.info
assays(1): reconstructed
rownames(13386): FO538757.2 AP006222.2 ... AC004556.1 AC240274.1
rowData names(1): rotation
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(1): batch
reducedDimNames(1): corrected
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<p>It is difficult to store the format of data fastMNN produces so we manipulate it to be of different class and call it back to Python.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o out2_all_genes
mat &lt;- assay(out_all_genes, &quot;reconstructed&quot;)
mat &lt;- as(mat, &quot;dgCMatrix&quot;)
out2_all_genes &lt;- SingleCellExperiment(list(counts = mat))
rownames(out2_all_genes) &lt;- rownames(out_all_genes)
colnames(out2_all_genes) &lt;- colnames(out_all_genes)
reducedDim(out2_all_genes) &lt;- reducedDim(out_all_genes)
reducedDimNames(out2_all_genes) &lt;- &#39;corrected_pca&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">out2_all_genes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obsm: &#39;corrected_pca&#39;
</pre></div>
</div>
</div>
</div>
<p>Note: As an output we get <code class="docutils literal notranslate"><span class="pre">out2_all_genes</span></code>, which has the corrected expression in <code class="docutils literal notranslate"><span class="pre">.X</span></code> and the corrected pca in <code class="docutils literal notranslate"><span class="pre">.obsm</span></code>. We can append these to our <code class="docutils literal notranslate"><span class="pre">adata</span></code> and work from there.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure that obs_names are in the same order in out2 and adata:</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="n">out2_all_genes</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not in right order&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure that var_names are in the same order in out2 and adata:</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="n">out2_all_genes</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not in right order&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;fastMNN_corrected&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out2_all_genes</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;fastMNN_pca&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">out2_all_genes</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;corrected_pca&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;, &#39;fastMNN_hvg&#39;, &#39;CellType_colors&#39;, &#39;fastMNN_all_genes&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;fastMNN_hvg_pca&#39;, &#39;X_umap_old&#39;, &#39;X_draw_graph_fa_old&#39;, &#39;fastMNN_pca&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;, &#39;fastMNN_hvg_corrected&#39;, &#39;fastMNN_corrected&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;, &#39;fastMNN_hvg_distances&#39;, &#39;fastMNN_hvg_connectivities&#39;, &#39;fastMNN_all_genes_distances&#39;, &#39;fastMNN_all_genes_connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="id3">
<h3>Nearest neighbors, UMAP, FDL<a class="headerlink" href="#id3" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;fastMNN_pca&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;fastMNN_all_genes&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;fastMNN_all_genes&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;fastMNN_all_genes&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id4">
<h3>Visualize the data<a class="headerlink" href="#id4" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of FastMNN corrected data (all genes)&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of un-corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of FastMNN corrected data (all genes)&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of un-corrected data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;FDL of un-corrected data&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/e5d31e2af8f8a96bf4a7d3482124c24c190e296bd01f51c7a5043cd053666c21.png"><img alt="_images/e5d31e2af8f8a96bf4a7d3482124c24c190e296bd01f51c7a5043cd053666c21.png" src="_images/e5d31e2af8f8a96bf4a7d3482124c24c190e296bd01f51c7a5043cd053666c21.png" style="width: 1014px; height: 763px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CellType&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/c647a2b44dd71e1605a5d5f0fdd03773e6cd12b61ce5431e3d2ad304f7626826.png"><img alt="_images/c647a2b44dd71e1605a5d5f0fdd03773e6cd12b61ce5431e3d2ad304f7626826.png" src="_images/c647a2b44dd71e1605a5d5f0fdd03773e6cd12b61ce5431e3d2ad304f7626826.png" style="width: 508px; height: 296px;" /></a>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="batch-effect-correction-seurat-integration">
<h1>Batch effect correction - Seurat Integration<a class="headerlink" href="#batch-effect-correction-seurat-integration" title="Link to this heading">#</a></h1>
<p>Seurat has an integrate method that allows for batch effect correction. This is also a very popular algorithm and operates by using mutual nearest neighbors to find common cells among batches and using them as anchors to correct for batch effect.</p>
<section id="implementation">
<h2>Implementation<a class="headerlink" href="#implementation" title="Link to this heading">#</a></h2>
<p>Note the input for Seurate Integrate is the same as for FastMNN.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;/data/peer/sharmar1/msk_workshop/workshop_2024_spring/session-4/batch_uncorrected_data.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">batch_id</span> <span class="o">=</span> <span class="s1">&#39;batch&#39;</span>
<span class="n">seurat_integrate_input</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> 
                 <span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;batch&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">j</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_id</span><span class="p">]]},</span> 
                                    <span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
                 <span class="n">var</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>In R, we need to convert SingleCellExperiment to a Seurat Object</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -i seurat_integrate_input
seurat_integrate_input
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 13386 15476 
metadata(0):
assays(1): X
rownames(13386): FO538757.2 AP006222.2 ... AC004556.1 AC240274.1
rowData names(0):
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(1): batch
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> 
logcounts(seurat_integrate_input) &lt;- assay(seurat_integrate_input, &quot;X&quot;)
print(&#39;logcounts done&#39;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;logcounts done&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
library(Seurat, quietly = TRUE)
seurat_data &lt;- as.Seurat(seurat_integrate_input, counts = &quot;logcounts&quot;, data = NULL)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    WARNING: The R package &quot;reticulate&quot; only fixed recently
    an issue that caused a segfault when used with rpy2:
    https://github.com/rstudio/reticulate/pull/1188
    Make sure that you use a version of that package that includes
    the fix.
    
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Attaching package: ‘sp’

The following object is masked from ‘package:IRanges’:

    %over%


Attaching package: ‘SeuratObject’

The following object is masked from ‘package:SummarizedExperiment’:

    Assays

The following object is masked from ‘package:GenomicRanges’:

    intersect

The following object is masked from ‘package:GenomeInfoDb’:

    intersect

The following object is masked from ‘package:IRanges’:

    intersect

The following object is masked from ‘package:S4Vectors’:

    intersect

The following object is masked from ‘package:BiocGenerics’:

    intersect

The following object is masked from ‘package:base’:

    intersect


Attaching package: ‘Seurat’

The following object is masked from ‘package:SummarizedExperiment’:

    Assays

Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)
Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
13386 features across 15476 samples within 1 assay 
Active assay: originalexp (13386 features, 0 variable features)
 2 layers present: counts, data
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
head(seurat_data@meta.data)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>                             orig.ident nCount_originalexp nFeature_originalexp
data_3p-AAACCTGAGCATCATC-0-0       data           1743.491                  865
data_3p-AAACCTGAGCTAGTGG-0-0       data           1925.536                 1314
data_3p-AAACCTGCACATTAGC-0-0       data           1678.488                  896
data_3p-AAACCTGCACTGTTAG-0-0       data           2180.531                 1519
data_3p-AAACCTGCATAGTAAG-0-0       data           2187.177                 1488
data_3p-AAACCTGCATGAACCT-0-0       data           2109.489                 1247
                             batch
data_3p-AAACCTGAGCATCATC-0-0     0
data_3p-AAACCTGAGCTAGTGG-0-0     0
data_3p-AAACCTGCACATTAGC-0-0     0
data_3p-AAACCTGCACTGTTAG-0-0     0
data_3p-AAACCTGCATAGTAAG-0-0     0
data_3p-AAACCTGCATGAACCT-0-0     0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# split the dataset into a list of seurat objects based on batches
data.list &lt;- SplitObject(seurat_data, split.by = &quot;batch&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# select features that are repeatedly variable across datasets for integration
# nfeatures referes to the number of highly variable genes
features &lt;- SelectIntegrationFeatures(object.list = data.list, nfeatures = 2000)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No variable features found for object1 in the object.list. Running FindVariableFeatures ...
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
No variable features found for object2 in the object.list. Running FindVariableFeatures ...
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat_data
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] FALSE
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
features_updated = c(features, &#39;DCN&#39;, &#39;PTPRC&#39;)
length(features_updated)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] 2002
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# find anchors to integrate using
seurat.anchors &lt;- FindIntegrationAnchors(object.list = data.list, anchor.features = features)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  |                                                  | 0 % ~calculating   |+++++++++++++++++++++++++                         | 50% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=02s  
  |                                                  | 0 % ~calculating   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=01m 44s
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Scaling features for provided objects
Finding all pairwise anchors
Running CCA
Merging objects
Finding neighborhoods
Finding anchors
	Found 17462 anchors
Filtering anchors
	Retained 6656 anchors
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o mat
# If you want to study the anchors, you can use:
head(seurat.anchors@anchors)
mat &lt;- as(seurat.anchors@anchors, &#39;data.frame&#39;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mat</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell1</th>
      <th>cell2</th>
      <th>score</th>
      <th>dataset1</th>
      <th>dataset2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>3.0</td>
      <td>47.0</td>
      <td>0.323529</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.0</td>
      <td>6429.0</td>
      <td>0.588235</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>5.0</td>
      <td>812.0</td>
      <td>0.470588</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>4655.0</td>
      <td>0.264706</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>8.0</td>
      <td>1977.0</td>
      <td>0.823529</td>
      <td>1</td>
      <td>2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>13308</th>
      <td>7201.0</td>
      <td>8092.0</td>
      <td>0.294118</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>13309</th>
      <td>7206.0</td>
      <td>8093.0</td>
      <td>0.264706</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>13310</th>
      <td>6469.0</td>
      <td>8093.0</td>
      <td>0.352941</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>13311</th>
      <td>758.0</td>
      <td>8095.0</td>
      <td>0.441176</td>
      <td>2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>13312</th>
      <td>463.0</td>
      <td>8097.0</td>
      <td>0.558824</td>
      <td>2</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>13312 rows × 5 columns</p>
</div></div></div>
</div>
</section>
<section id="integrate">
<h2>Integrate<a class="headerlink" href="#integrate" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# Integrate:
seurat.combined &lt;- IntegrateData(anchorset = seurat.anchors)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Merging dataset 2 into 1
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Integrating data
</pre></div>
</div>
</div>
</div>
</section>
<section id="study-the-output">
<h2>Study the output<a class="headerlink" href="#study-the-output" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat.combined
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
15386 features across 15476 samples within 2 assays 
Active assay: integrated (2000 features, 2000 variable features)
 1 layer present: data
 1 other assay present: originalexp
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
# convert to SingleCellExperiment object so we can export to Python
seurat_integrated &lt;- as.SingleCellExperiment(seurat.combined)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span>
seurat_integrated
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 2000 15476 
metadata(0):
assays(1): logcounts
rownames(2000): LYZ S100A8 ... RP1-151F17.2 U47924.31
rowData names(0):
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(5): orig.ident nCount_originalexp nFeature_originalexp
  batch ident
reducedDimNames(0):
mainExpName: integrated
altExpNames(1): originalexp
</pre></div>
</div>
</div>
</div>
<p>We note that Seurat Integrate selects for highly variable genes, so the output only has <code class="docutils literal notranslate"><span class="pre">nFeatures</span></code>, in which we specified to be 2000 (which you can change) above. So Seurat integrate only corrects for the those selected genes, therefore allows only restricted downstream analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">R</span> -o seurat_integrated
seurat_integrated
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment 
dim: 2000 15476 
metadata(0):
assays(1): logcounts
rownames(2000): LYZ S100A8 ... RP1-151F17.2 U47924.31
rowData names(0):
colnames(15476): data_3p-AAACCTGAGCATCATC-0-0
  data_3p-AAACCTGAGCTAGTGG-0-0 ... data_5p-TTTGTCATCTTACCGC-1-1
  data_5p-TTTGTCATCTTGTTTG-1-1
colData names(5): orig.ident nCount_originalexp nFeature_originalexp
  batch ident
reducedDimNames(0):
mainExpName: integrated
altExpNames(1): originalexp
</pre></div>
</div>
</div>
</div>
<p>Look at it in Python</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seurat_integrated</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 2000
    obs: &#39;orig.ident&#39;, &#39;nCount_originalexp&#39;, &#39;nFeature_originalexp&#39;, &#39;batch&#39;, &#39;ident&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="add-to-adata">
<h2>Add to adata<a class="headerlink" href="#add-to-adata" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure that obs_names are in the same order in out2 and adata:</span>
<span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">item</span> <span class="o">!=</span> <span class="n">seurat_integrated</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not in right order&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We store the integrated data in `.obsm` and the corresponding genes in `.uns`.</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;seurat_integrated_data&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seurat_integrated</span><span class="o">.</span><span class="n">X</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;seurat_integrated_features&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seurat_integrated</span><span class="o">.</span><span class="n">var_names</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;, &#39;fastMNN_hvg&#39;, &#39;CellType_colors&#39;, &#39;fastMNN_all_genes&#39;, &#39;seurat_integrated_features&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;fastMNN_hvg_pca&#39;, &#39;X_umap_old&#39;, &#39;X_draw_graph_fa_old&#39;, &#39;fastMNN_pca&#39;, &#39;seurat_integrated_data&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;, &#39;fastMNN_hvg_corrected&#39;, &#39;fastMNN_corrected&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;, &#39;fastMNN_hvg_distances&#39;, &#39;fastMNN_hvg_connectivities&#39;, &#39;fastMNN_all_genes_distances&#39;, &#39;fastMNN_all_genes_connectivities&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="pca-nearest-neighbors-umap-fdl">
<h2>PCA, Nearest Neighbors, UMAP, FDL<a class="headerlink" href="#pca-nearest-neighbors-umap-fdl" title="Link to this heading">#</a></h2>
<p>Since the Seurat results are stored in obsm, we run PCA on it only:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seurat_pca</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;seurat_integrated_data&#39;</span><span class="p">],</span> <span class="n">n_comps</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">return_info</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seurat_pca</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(15476, 30)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_pca_seurat&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seurat_pca</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Nearest neighbors on PCA</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca_seurat&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> 
                <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;seurat&#39;</span><span class="p">)</span>

<span class="c1"># UMAP, FDL using the nearest neighbors</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span> <span class="o">=</span> <span class="s1">&#39;seurat&#39;</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">draw_graph</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s1">&#39;fa&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;seurat&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id5">
<h2>Visualize<a class="headerlink" href="#id5" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of Seurat corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;UMAP of un-corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of Seurat corrected data&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">c0</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
<span class="n">c1</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-0&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_draw_graph_fa_old&#39;</span><span class="p">][</span><span class="n">c1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Batch-1&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;FDL of un-corrected data&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;FDL of un-corrected data&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/1bc44b9283fda94007adccdc230da1b07344b3c5100115f0a8acdbf3dfe7a732.png"><img alt="_images/1bc44b9283fda94007adccdc230da1b07344b3c5100115f0a8acdbf3dfe7a732.png" src="_images/1bc44b9283fda94007adccdc230da1b07344b3c5100115f0a8acdbf3dfe7a732.png" style="width: 1014px; height: 763px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CellType&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/2f4e6e60c7eae073b64300f399642136b9e7a3a234726ab6a6054a5db5718c20.png"><img alt="_images/2f4e6e60c7eae073b64300f399642136b9e7a3a234726ab6a6054a5db5718c20.png" src="_images/2f4e6e60c7eae073b64300f399642136b9e7a3a234726ab6a6054a5db5718c20.png" style="width: 508px; height: 296px;" /></a>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="ideas-to-quantifying-effectiveness-of-batch-effect-correction">
<h1>Ideas to quantifying effectiveness of batch effect correction<a class="headerlink" href="#ideas-to-quantifying-effectiveness-of-batch-effect-correction" title="Link to this heading">#</a></h1>
<p>Aside from visually comparing the UMAPs (with cell type/gene expression information) before and after batch correction, we can also quantify how effective our batch correction is using <code class="docutils literal notranslate"><span class="pre">batch</span> <span class="pre">entropy</span></code>.</p>
<p>At a high level, batch entropy is an indicator of the degree of mixture between cells of different batches. To illustrate, for a given cell X, we first compute the cells in its nearest neighbors and identify their batch ids. If this neighborhood around X is well-mixed with cells from different batches then it means that the cells similar to X are present in many different batches and that the batches align well. This would imply that the more heterogeneous (in terms of batch id) the neighborhood of each cell is, the less batch effect there is. We quantify this notion of heterogeneity using entropy.</p>
<p>Computationally:</p>
<ul class="simple">
<li><p>Given a cell <span class="math notranslate nohighlight">\(i\)</span>, we first compute <span class="math notranslate nohighlight">\(k\)</span> nearest neighbors around it and their batch id</p></li>
<li><p>We compute the fraction (out of <span class="math notranslate nohighlight">\(k\)</span>) of cells from each batch id in that neighborhood</p></li>
<li><p>This represents the probability vector of batch distribution around cell <span class="math notranslate nohighlight">\(i\)</span> (Call it <span class="math notranslate nohighlight">\(p_i\)</span>)</p></li>
<li><p>We repeat this for all cells, one-by-one</p></li>
<li><p>We then calculate the Shannon entropy: $<span class="math notranslate nohighlight">\(H = -\sum_{i=1}^N p_i \log_2(p_i),\)</span>$ where N is the total number of cells.</p></li>
</ul>
<p>Thus, if there are obvious batch effects, we expect cells of the same batch to be near each other (more homogenous) distant from other batches, so the overall batch entropy distribution across all cells will be lower. Similarly, after effective batch correction, we expect the cells to be more mixed (more heterogenous), and the overall batch entropy distribution across cells to be higher.</p>
<p>Like always, we will compute the nearest neighbors on PCA space.</p>
<section id="before-correction">
<h2>Before correction<a class="headerlink" href="#before-correction" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;Before Batch Correction&#39;</span><span class="p">,</span> <span class="n">basis</span> <span class="o">=</span> <span class="s1">&#39;umap_old&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/38e5be33940e478d924bf90ddfeb2f71a1833516353ffcff61fded0325dae12c.png"><img alt="_images/38e5be33940e478d924bf90ddfeb2f71a1833516353ffcff61fded0325dae12c.png" src="_images/38e5be33940e478d924bf90ddfeb2f71a1833516353ffcff61fded0325dae12c.png" style="width: 324px; height: 319px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 15476 × 13386
    obs: &#39;Sample&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;batch&#39;, &#39;louvain&#39;, &#39;anno&#39;, &#39;Method&#39;, &#39;CellType&#39;, &#39;batch_id&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;pheno_leiden&#39;
    var: &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;hvg&#39;, &#39;id_hvg&#39;, &#39;pca&#39;, &#39;neighbors_30&#39;, &#39;umap&#39;, &#39;draw_graph&#39;, &#39;pheno_jaccard_q&#39;, &#39;batch_id_colors&#39;, &#39;pheno_leiden_colors&#39;, &#39;fastMNN_hvg&#39;, &#39;CellType_colors&#39;, &#39;fastMNN_all_genes&#39;, &#39;seurat_integrated_features&#39;, &#39;seurat&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;X_draw_graph_fa&#39;, &#39;fastMNN_hvg_pca&#39;, &#39;X_umap_old&#39;, &#39;X_draw_graph_fa_old&#39;, &#39;fastMNN_pca&#39;, &#39;seurat_integrated_data&#39;, &#39;X_pca_seurat&#39;
    varm: &#39;PCs&#39;
    layers: &#39;raw_counts&#39;, &#39;norm_counts&#39;, &#39;fastMNN_hvg_corrected&#39;, &#39;fastMNN_corrected&#39;
    obsp: &#39;neighbors_30_distances&#39;, &#39;neighbors_30_connectivities&#39;, &#39;pheno_jaccard_ig&#39;, &#39;fastMNN_hvg_distances&#39;, &#39;fastMNN_hvg_connectivities&#39;, &#39;fastMNN_all_genes_distances&#39;, &#39;fastMNN_all_genes_connectivities&#39;, &#39;seurat_distances&#39;, &#39;seurat_connectivities&#39;
</pre></div>
</div>
</div>
</div>
<p>Since we already have computed the nearest neighbors (<code class="docutils literal notranslate"><span class="pre">adata.obsp['neighbors_30_distances']</span></code>), we simply need to get the batch-id of each cell in each of the neighborhoods and compute entropy. We can use the following function to compute entropy:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scipy</span>
<span class="c1"># a function to compute entropy for a cell given the batch labels of cells in its neighborhood</span>
<span class="k">def</span> <span class="nf">compute_entropy</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">value</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">scipy</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">entropy</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">base</span><span class="o">=</span><span class="n">base</span><span class="p">)</span>

<span class="c1"># a function to convert scanpy&#39;s nearest neighbor output to a more standard matrix </span>
<span class="k">def</span> <span class="nf">create_nn_mat</span><span class="p">(</span><span class="n">adata_temp</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>
    <span class="p">[</span><span class="n">row</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">val</span><span class="p">]</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">adata_temp</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
    <span class="n">nn_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="n">adata_temp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">sum</span><span class="p">(</span><span class="n">row</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata_temp</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">nn_mat</span><span class="p">[</span><span class="n">item</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">col</span><span class="p">[</span><span class="n">row</span> <span class="o">==</span> <span class="n">item</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">nn_mat</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># convert scanpy&#39;s nearest neighbor distance to a standard matrix</span>
<span class="n">nn_mat</span> <span class="o">=</span> <span class="n">create_nn_mat</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;neighbors_30_distances&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute entropy before correction</span>
<span class="n">entropy_before</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nn_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">nn_mat</span><span class="p">[</span><span class="n">r</span><span class="p">]]</span>
    <span class="n">entropy_val</span> <span class="o">=</span> <span class="n">compute_entropy</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">base</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">entropy_before</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">entropy_val</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="after-correction">
<h2>After correction<a class="headerlink" href="#after-correction" title="Link to this heading">#</a></h2>
<p>For illustration, let’s take Seurat integrated as an example of after batch effect correction:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">],</span> <span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;After Batch Correction, using Seurat&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/sharmar1/miniconda3/envs/workshop_2024/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/8d1df5b73d6fbec7dfea3719a4e130f42fa8ad5fadd864b9702ba4205dcd9fb2.png"><img alt="_images/8d1df5b73d6fbec7dfea3719a4e130f42fa8ad5fadd864b9702ba4205dcd9fb2.png" src="_images/8d1df5b73d6fbec7dfea3719a4e130f42fa8ad5fadd864b9702ba4205dcd9fb2.png" style="width: 325px; height: 296px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">nn_mat_after</span> <span class="o">=</span> <span class="n">create_nn_mat</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key</span> <span class="o">=</span> <span class="s1">&#39;seurat_distances&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">entropy_after</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nn_mat_after</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="n">entropy_after</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">compute_entropy</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">nn_mat_after</span><span class="p">[</span><span class="n">r</span><span class="p">]],</span> <span class="n">base</span> <span class="o">=</span> <span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">entropy_after</span><span class="p">,</span> <span class="mi">10</span><span class="p">);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Entropy&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;After FastMNN correction&#39;</span><span class="p">)</span>


<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">entropy_before</span><span class="p">,</span> <span class="mi">10</span><span class="p">);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Entropy&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Before correction&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;Before correction&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/2018b66245766161d5689bb21ca45f2cb8c819b115882d586e66bbea48954e04.png"><img alt="_images/2018b66245766161d5689bb21ca45f2cb8c819b115882d586e66bbea48954e04.png" src="_images/2018b66245766161d5689bb21ca45f2cb8c819b115882d586e66bbea48954e04.png" style="width: 1081px; height: 440px;" /></a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">8</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="n">entropy_after</span><span class="p">,</span> 
                 <span class="n">vmin</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mi">1</span><span class="p">);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;After Seurat correction&#39;</span><span class="p">)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">im1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][:,</span> <span class="mi">0</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap_old&#39;</span><span class="p">][:,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">s</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="n">entropy_before</span><span class="p">,</span> 
                 <span class="n">vmin</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mi">1</span><span class="p">);</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">im1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Before correction&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;Before correction&#39;)
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/647f2243fca6e17a6c85314fc3aeffc9b5d1bcabb02db5b43333e55c9a71ff3f.png"><img alt="_images/647f2243fca6e17a6c85314fc3aeffc9b5d1bcabb02db5b43333e55c9a71ff3f.png" src="_images/647f2243fca6e17a6c85314fc3aeffc9b5d1bcabb02db5b43333e55c9a71ff3f.png" style="width: 997px; height: 407px;" /></a>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "conda-env-workshop_2024-py"
        },
        kernelOptions: {
            name: "conda-env-workshop_2024-py",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'conda-env-workshop_2024-py'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Introduction</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#load-packages">Load packages</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-multiple-samples-into-scanpy">Loading multiple samples into Scanpy</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-1">batch-1</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-2">batch-2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#concatenate">Concatenate</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#basic-qc-and-normalization">Basic QC and normalization</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#highly-variable-genes-pca-nearest-neighbor-graph-umap-fdl-phenograph">Highly variable genes, PCA, Nearest Neighbor Graph, UMAP, FDL, Phenograph</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-results">Visualize the results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-some-genes">Visualize some genes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#highlight-b-cells">Highlight B cells</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction">Batch Effect Correction</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-scanorama">Batch effect correction - Scanorama</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#nearest-neighbors-umap-fdl">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize">Visualize</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#true-celltypes">True celltypes</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-harmony">Batch effect correction - Harmony</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-fastmnn">Batch effect correction - FastMNN</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-with-hvgs">Implementation with HVGs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#process-the-corrected-output">Process the corrected output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualize-the-data">Visualize the data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation-with-all-genes">Implementation with all genes</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Process the corrected output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Nearest neighbors, UMAP, FDL</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Visualize the data</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-effect-correction-seurat-integration">Batch effect correction - Seurat Integration</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#implementation">Implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#integrate">Integrate</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#study-the-output">Study the output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-to-adata">Add to adata</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-nearest-neighbors-umap-fdl">PCA, Nearest Neighbors, UMAP, FDL</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Visualize</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#ideas-to-quantifying-effectiveness-of-batch-effect-correction">Ideas to quantifying effectiveness of batch effect correction</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#before-correction">Before correction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#after-correction">After correction</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sara Jimenez
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=26a4bc78f4c0ddb94549"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=26a4bc78f4c0ddb94549"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>