# GSCN workshop 2024

This repository is a collection of tutorials presented during the [GSCN workshop 2024 ‘Computational single cell genomics for stem cell biologists’](https://www.gscn.org/conferences-events/gscn-events/working-group-workshops) - November 7-8, 2024.  

The corresponding Jupyter Book is rendered [here](https://sarajimenez.github.io/GSCN_workshop_2024/).

## Table of Contents

- [Material](#Material)
- [Session-1](#Session-1)
- [Session-2](#Session-2)
- [Session-3](#Session-3)
- [Session-4](#Session-4)

## Material

The data for each session can be found [here](https://drive.google.com/drive/folders/1mD36tCwFMJ5g4Fh5p5nptDXy_ObmK7ww?usp=sharing).  
Material for the workshop has been adapted from the [single-cell best practices book](https://www.sc-best-practices.org/preamble.html), a live project from the Theis lab.   
Recordings of previous single-cell RNA seq data analysis workshops can be found [here](https://www.youtube.com/playlist?list=PL4rcQcNPLZxWQQH7LlRBMkAo5NWuHX1e3).


## Session-1 

1. Preprocessing and Quality Control (QC)
2. Data normalization

## Session-2

1. Dimensionality reduction 
2. Clustering and annotation 
3. DGE

## Session-3

1. Trajectory analysis
2. Multisample analysis 

## Session-4

1. Introduction to spatial transcriptomics

## Acknowledments

Special thanks to the scverse community (especially the developers and the maintainers of scanpy and squidpy packages) and all the Theis lab for valuable contributions.   
Huge thanks also go to [Roshan Sharma](https://www.mskcc.org/research/ski/labs/members/roshan-sharma) from Dana Pe'er's lab for kindly providing material for the workshop.
